{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import path, makedirs\n",
    "import glob\n",
    "import json\n",
    "from jsonschema import validate, Draft7Validator\n",
    "from tempfile import TemporaryDirectory\n",
    "from Bio import SeqIO, Entrez\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Alphabet import generic_protein\n",
    "from sys import exit\n",
    "import datetime\n",
    "import time\n",
    "import re\n",
    "import urllib3\n",
    "import certifi\n",
    "http = urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=certifi.where())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gbk(nucl_acc, email, clean_cache = False):\n",
    "    mibig_gbks_folder = \"../../preprocessed/gbks_not_in_ncbi/\"\n",
    "    cache_folder = \"../../preprocessed/cached_gbks/\"\n",
    "    cache_path = path.join(cache_folder, \"{}.gbk\".format(nucl_acc))\n",
    "\n",
    "    if clean_cache or not path.exists(cache_path) or path.getsize(cache_path) < 100:\n",
    "        if re.match(\"^MIBIG:BGC\\d{7}\\.\\d+$\", nucl_acc): # fetch from gbk folder\n",
    "            gbk_path = path.join(mibig_gbks_folder, \"{}.gbk\".format(nucl_acc.split(\":\")[1]))\n",
    "            return open(gbk_path, \"r\")\n",
    "        else: # (re)download from ncbi\n",
    "            Entrez.email = email\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=nucl_acc, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "            if not path.exists(cache_folder):\n",
    "                makedirs(cache_folder)\n",
    "        with open(cache_path, \"w\") as gbk_file:\n",
    "            gbk_file.write(handle.read())\n",
    "            \n",
    "    return open(cache_path, \"r\")\n",
    "\n",
    "# from antismash\n",
    "def get_aa_translation(seq_record, feature):\n",
    "    \"\"\"Obtain content for translation qualifier for specific CDS feature in sequence record\"\"\"\n",
    "    extracted = feature.extract(seq_record.seq).ungap('-')\n",
    "\n",
    "    # ensure the extracted section is a multiple of three by trimming any excess\n",
    "    if len(extracted) % 3 != 0:\n",
    "        extracted = extracted[:-(len(extracted) % 3)]\n",
    "\n",
    "    fasta_seq = extracted.translate(to_stop=True)\n",
    "    if len(fasta_seq) == 0:\n",
    "        print(\"Retranslating {} with stop codons\".format(feature.id))\n",
    "        fasta_seq = extracted.translate()\n",
    "\n",
    "    # replace ambiguous aminos with an explicit unknown\n",
    "    string_version = str(fasta_seq)\n",
    "    for bad in \"*BJOUZ\":\n",
    "        string_version = string_version.replace(bad, \"X\")\n",
    "\n",
    "    # and remove any gaps\n",
    "    string_version = string_version.replace(\"-\", \"\")\n",
    "    fasta_seq = Seq(string_version, generic_protein)\n",
    "\n",
    "    return fasta_seq\n",
    "\n",
    "def get_aa_sequence(feature, to_stop=False):\n",
    "    \"\"\"Extract sequence from specific CDS feature in sequence record\"\"\"\n",
    "    fasta_seq = feature.qualifiers['translation'][0]\n",
    "    if \"*\" in fasta_seq:\n",
    "        if to_stop:\n",
    "            fasta_seq = fasta_seq.split('*')[0]\n",
    "        else:\n",
    "            fasta_seq = fasta_seq.replace(\"*\",\"X\")\n",
    "    if \"-\" in fasta_seq:\n",
    "        fasta_seq = fasta_seq.replace(\"-\",\"\")\n",
    "    return fasta_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_mibig_final_gbk(bgc_id, clean_cache = False):\n",
    "    cache_folder = \"../../preprocessed/cached_mibig_finalgbks/\"\n",
    "    cache_path = path.join(cache_folder, \"{}.1.final.gbk\".format(bgc_id))\n",
    "    \n",
    "    if clean_cache or not path.exists(cache_path) or path.getsize(cache_path) < 100:\n",
    "        mibig_url = \"https://mibig.secondarymetabolites.org/repository/{}/{}.1.final.gbk\".format(bgc_id, bgc_id)\n",
    "        resp = http.request('GET', mibig_url)\n",
    "        with open(cache_path, \"w\") as cf:\n",
    "            cf.write(resp.data.decode('utf-8', 'ignore'))\n",
    "        \n",
    "    return open(cache_path, \"r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../../preprocessed/gene_counts.tsv\", \"w\") as o:\n",
    "    o.write(\"mibig_id\\tnucl_acc\\tloc\\tfrom_gbk\\tfrom_genefinding\\tnew_annot\\tre_annot\\n\")\n",
    "\n",
    "with open(\"../../preprocessed/.tsv\", \"w\") as o:\n",
    "    o.write(\"\")\n",
    "    \n",
    "def fetch_additional_genes(annotation_data):\n",
    "    # fetch seq records from ncbi (or use cached files)\n",
    "    nucl_acc = annotation_data[\"general_params\"][\"loci\"][\"accession\"]\n",
    "    #if re.match(\"^MIBIG:BGC\\d{7}\\.\\d+$\", nucl_acc.upper()): # using mibig accessions\n",
    "    #    return\n",
    "    nucl_acc_to_downloads = [nucl_acc]\n",
    "\n",
    "    # fetch sequence gbks\n",
    "    email = annotation_data[\"personal\"][\"submitter_email\"] # probably not a good idea, but let's just put it as it is for now\n",
    "    nucl_records = {}\n",
    "    for nucl_acc_to_download in nucl_acc_to_downloads:\n",
    "        num_try = 1\n",
    "        clean_cache = False\n",
    "        while num_try < 6:\n",
    "            try:\n",
    "                with fetch_gbk(nucl_acc_to_download, email, clean_cache) as gbk_handle:\n",
    "                    seq_record = SeqIO.read(gbk_handle, \"genbank\") # the gbk should contains only 1 file\n",
    "                    if len(seq_record.seq) < 1:\n",
    "                        raise Exception(\"Empty sequence record {}\".format(nucl_acc_to_download))\n",
    "                    nucl_records[nucl_acc_to_download] = seq_record\n",
    "                    break\n",
    "            except:\n",
    "                print(\"Error...\")\n",
    "                clean_cache = True\n",
    "            num_try += 1\n",
    "            time.sleep(5)\n",
    "        if nucl_acc_to_download not in nucl_records: # failed to download NCBI data\n",
    "            print(\"{} Failed to download: http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id={}&rettype=gbwithparts&retmode=text\".format(annotation_data[\"general_params\"][\"mibig_accession\"], nucl_acc_to_download))\n",
    "            return {}\n",
    "\n",
    "    # fetch finalgbks\n",
    "    bgc_id = annotation_data[\"general_params\"][\"mibig_accession\"]\n",
    "    mibig_genes = {}\n",
    "    with fetch_mibig_final_gbk(bgc_id) as mibig_gbk_handle:\n",
    "        mibig_seq_records = SeqIO.parse(mibig_gbk_handle, \"genbank\")\n",
    "        for mibig_seq_record in mibig_seq_records:\n",
    "            match = re.search(\"(between(?P<start>\\d+)-(?P<end>\\d+)ntfrom){0,1}GenBankID(?P<acc>[A-Z0-9_\\.]+)\\.\", mibig_seq_record.annotations['comment'].replace(\" \", \"\").replace(\"\\n\", \"\"))\n",
    "            if match:\n",
    "                nucl_accession = match.group(\"acc\").upper()\n",
    "                if match.group(\"start\"):\n",
    "                    cl_start = int(match.group(\"start\"))\n",
    "                    cl_end = int(match.group(\"end\"))\n",
    "                    if nucl_accession not in mibig_genes:\n",
    "                        mibig_genes[nucl_accession] = []\n",
    "                    for feature in mibig_seq_record.features:\n",
    "                        if feature.type == \"CDS\":\n",
    "                            f_start = cl_start - 1 + feature.location.start\n",
    "                            f_end = cl_start - 1 + feature.location.end\n",
    "                            feature.location = FeatureLocation(f_start, f_end)\n",
    "                            mibig_genes[nucl_accession].append(feature)\n",
    "                else:\n",
    "                    mibig_genes[nucl_accession] = [feature for feature in mibig_seq_record.features if feature.type == \"CDS\"]\n",
    "            else:\n",
    "                print(mibig_seq_record.annotations['comment'])\n",
    "        \n",
    "    # compare with mibig clustergbk, then\n",
    "    #for loci in annotation_data[\"general_params\"][\"loci\"][\"nucl_acc\"]:\n",
    "    loci = annotation_data[\"general_params\"][\"loci\"]\n",
    "    if True:\n",
    "        gene_counts = 0\n",
    "        seq_record = nucl_records[loci[\"accession\"].upper()]        \n",
    "        record = {\n",
    "            \"nucl_acc\": loci[\"accession\"],\n",
    "            \"start_pos\": loci[\"start_coord\"] if \"start_coord\" in loci and loci[\"start_coord\"] > 0 else 1, # use genbank-offset\n",
    "            \"end_pos\": loci[\"end_coord\"] if \"end_coord\" in loci and loci[\"end_coord\"] > 0 else len(seq_record.seq) # use genbank-offset\n",
    "        }\n",
    "        genes_in_gbk = {}\n",
    "        for feature in seq_record.features:\n",
    "            if feature.type == \"CDS\" and feature.location.start >= record[\"start_pos\"]-1 and feature.location.end <= record[\"end_pos\"]:\n",
    "                new_gene = {}\n",
    "                # name\n",
    "                if \"gene\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_name\"] = feature.qualifiers[\"gene\"][0]\n",
    "                elif \"protein_id\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_name\"] = feature.qualifiers[\"protein_id\"][0]\n",
    "                elif \"locus_tag\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_name\"] = feature.qualifiers[\"locus_tag\"][0]\n",
    "                # id\n",
    "                if \"locus_tag\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_id\"] = feature.qualifiers[\"locus_tag\"][0]\n",
    "                elif \"protein_id\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_id\"] = feature.qualifiers[\"protein_id\"][0]\n",
    "                elif \"gene\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_id\"] = feature.qualifiers[\"gene\"][0]\n",
    "                # holder for everything (for matching purpose only)\n",
    "                new_gene[\"ids\"] = []\n",
    "                if \"locus_tag\" in feature.qualifiers:\n",
    "                    new_gene[\"ids\"].append(feature.qualifiers[\"locus_tag\"][0].lower())\n",
    "                if \"protein_id\" in feature.qualifiers:\n",
    "                    new_gene[\"ids\"].append(feature.qualifiers[\"protein_id\"][0].lower())\n",
    "                if \"gene\" in feature.qualifiers:\n",
    "                    new_gene[\"ids\"].append(feature.qualifiers[\"gene\"][0].lower())\n",
    "                # locs\n",
    "                new_gene[\"gene_startpos\"] = feature.location.start + 1\n",
    "                new_gene[\"gene_endpos\"] = feature.location.end\n",
    "                # strand\n",
    "                new_gene[\"strand\"] = feature.location.strand\n",
    "                # annotation\n",
    "                if \"product\" in feature.qualifiers:\n",
    "                    new_gene[\"gene_annotation\"] = feature.qualifiers[\"product\"][0]\n",
    "                # set aa sequence\n",
    "                if \"translation\" in feature.qualifiers:\n",
    "                    new_gene[\"aa_seq\"] = str(get_aa_sequence(feature, False))\n",
    "                else: # translate aa sequence from nucleotide sequence\n",
    "                    new_gene[\"aa_seq\"] = str(get_aa_translation(seq_record, feature))\n",
    "                # dummy -- todo: set this based on ..?\n",
    "                new_gene[\"gene_function\"] = \"Unknown\"\n",
    "                gene = (feature.location.start, feature.location.end) # use locus as identifier\n",
    "                genes_in_gbk[gene] = new_gene\n",
    "        \n",
    "        genes_from_mibig = {}\n",
    "        if len(genes_in_gbk.keys()) < 1: # fetch genefinder's genes from mibig\n",
    "            mibig_features = mibig_genes[loci[\"accession\"].upper()]\n",
    "            for feature in mibig_features:\n",
    "                gene = (feature.location.start, feature.location.end) # use locus as identifier\n",
    "                if feature.type == \"CDS\" and feature.location.start >= record[\"start_pos\"]-1 and feature.location.end <= record[\"end_pos\"]:\n",
    "                    if gene not in genes_in_gbk:\n",
    "                        genes_from_mibig[gene] = feature\n",
    "                        ## append to annotation json\n",
    "                        ## todo: check existing annotations\n",
    "                        if \"genes\" not in annotation_data[\"general_params\"]:\n",
    "                            annotation_data[\"general_params\"][\"genes\"] = {\"gene\": []}\n",
    "                        elif \"gene\" not in annotation_data[\"general_params\"][\"genes\"]:\n",
    "                            annotation_data[\"general_params\"][\"genes\"][\"gene\"] = []\n",
    "                        new_gene = {}\n",
    "                        # name\n",
    "                        if \"gene\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_name\"] = feature.qualifiers[\"gene\"][0]\n",
    "                        elif \"protein_id\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_name\"] = feature.qualifiers[\"protein_id\"][0]\n",
    "                        elif \"locus_tag\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_name\"] = feature.qualifiers[\"locus_tag\"][0]\n",
    "                        # id\n",
    "                        if \"locus_tag\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_id\"] = feature.qualifiers[\"locus_tag\"][0]\n",
    "                        elif \"protein_id\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_id\"] = feature.qualifiers[\"protein_id\"][0]\n",
    "                        elif \"gene\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_id\"] = feature.qualifiers[\"gene\"][0]\n",
    "                        # locs\n",
    "                        new_gene[\"gene_startpos\"] = feature.location.start + 1\n",
    "                        new_gene[\"gene_endpos\"] = feature.location.end\n",
    "                        # strand\n",
    "                        new_gene[\"strand\"] = feature.location.strand\n",
    "                        # annotation\n",
    "                        if \"product\" in feature.qualifiers:\n",
    "                            new_gene[\"gene_annotation\"] = feature.qualifiers[\"product\"][0]\n",
    "                        # set aa sequence\n",
    "                        if \"translation\" in feature.qualifiers:\n",
    "                            new_gene[\"aa_seq\"] = str(get_aa_sequence(feature, False))\n",
    "                        else: # translate aa sequence from nucleotide sequence\n",
    "                            new_gene[\"aa_seq\"] = str(get_aa_translation(seq_record, feature))\n",
    "                        # dummy -- todo: set this based on ..?\n",
    "                        new_gene[\"gene_function\"] = \"Unknown\"\n",
    "                        # update if exist, add if not\n",
    "                        is_updated = False                        \n",
    "                        for annot_gene in annotation_data[\"general_params\"][\"genes\"][\"gene\"]:\n",
    "                            if \"gene_id\" in annot_gene and new_gene[\"gene_id\"] == annot_gene[\"gene_id\"]:\n",
    "                                annot_gene[\"gene_startpos\"] = new_gene[\"gene_startpos\"]\n",
    "                                annot_gene[\"gene_endpos\"] = new_gene[\"gene_endpos\"]\n",
    "                                if \"gene_name\" not in annot_gene and \"gene_name\" in new_gene:\n",
    "                                    annot_gene[\"gene_name\"] = new_gene[\"gene_name\"]\n",
    "                                if \"gene_annotation\" not in annot_gene and \"gene_annotation\" in new_gene:\n",
    "                                    annot_gene[\"gene_annotation\"] = new_gene[\"gene_annotation\"]\n",
    "                                if \"aa_seq\" not in annot_gene and \"aa_seq\" in new_gene:\n",
    "                                    annot_gene[\"aa_seq\"] = new_gene[\"aa_seq\"]\n",
    "                        if not is_updated:\n",
    "                            annotation_data[\"general_params\"][\"genes\"][\"gene\"].append(new_gene)\n",
    "        \n",
    "        # validate annotated genes\n",
    "        reannotation_count = 0\n",
    "        newannotation_count = 0\n",
    "        newannotated_genes = []\n",
    "        annot_genes = []\n",
    "        if \"genes\" in annotation_data[\"general_params\"] and \"gene\" in annotation_data[\"general_params\"][\"genes\"]:\n",
    "            for annot_gene in annotation_data[\"general_params\"][\"genes\"][\"gene\"]:\n",
    "                corresponding_gbk_gene = None\n",
    "                for key in genes_in_gbk:\n",
    "                    gbk_gene = genes_in_gbk[key]\n",
    "                    if ((\"gene_id\" in annot_gene and annot_gene[\"gene_id\"].lower() in gbk_gene[\"ids\"])\n",
    "                        or (\"gene_name\" in annot_gene and annot_gene[\"gene_name\"].lower() in gbk_gene[\"ids\"])):\n",
    "                        corresponding_gbk_gene = gbk_gene\n",
    "                        break\n",
    "                gbk_gene = corresponding_gbk_gene\n",
    "                if gbk_gene != None:\n",
    "                    if \"gene_id\" in gbk_gene:\n",
    "                        annot_gene[\"gene_id\"] = gbk_gene[\"gene_id\"]\n",
    "                    if \"gene_name\" in gbk_gene:\n",
    "                        annot_gene[\"gene_name\"] = gbk_gene[\"gene_name\"]\n",
    "                    # check if gene_id or gene_name is in gbk, then discard positional information (?)\n",
    "                    if \"gene_startpos\" in annot_gene:\n",
    "                        del annot_gene[\"gene_startpos\"]\n",
    "                    if \"gene_endpos\" in annot_gene:\n",
    "                        del annot_gene[\"gene_endpos\"]\n",
    "                    if \"strand\" in annot_gene:\n",
    "                        del annot_gene[\"strand\"]\n",
    "                    # check if there is any point on putting the annotation\n",
    "                    if \"gene_annotation\" in annot_gene and (\"gene_annotation\" in gbk_gene and annot_gene[\"gene_annotation\"].lower() != gbk_gene[\"gene_annotation\"].lower()) or (\"gene_annotation\" not in gbk_gene):\n",
    "                        annot_genes.append(annot_gene)\n",
    "                        reannotation_count += 1\n",
    "                    elif \"gene_function\" in annot_gene and (\"gene_function\" in gbk_gene and annot_gene[\"gene_function\"].lower() != gbk_gene[\"gene_function\"].lower()) or (\"gene_function\" not in gbk_gene):\n",
    "                        annot_genes.append(annot_gene)\n",
    "                        reannotation_count += 1\n",
    "                    elif \"aa_seq\" in annot_gene and (\"aa_seq\" in gbk_gene and annot_gene[\"aa_seq\"] != gbk_gene[\"aa_seq\"]) or (\"aa_seq\" not in gbk_gene):\n",
    "                        annot_genes.append(annot_gene)\n",
    "                        reannotation_count += 1\n",
    "                    else:\n",
    "                        for key in annot_gene:\n",
    "                            if key not in gbk_gene:\n",
    "                                annot_genes.append(annot_gene)\n",
    "                                reannotation_count += 1\n",
    "                                break\n",
    "                else:\n",
    "                    # check start&endpos, if not within range of record then discard positional information\n",
    "                    # if startpos = endpos, also discard positional information\n",
    "                    if \"gene_startpos\" in annot_gene:\n",
    "                        if \"gene_endpos\" in annot_gene:\n",
    "                            if ((annot_gene[\"gene_startpos\"] < record[\"start_pos\"]) or (annot_gene[\"gene_endpos\"] > record[\"end_pos\"])\n",
    "                               or (annot_gene[\"gene_startpos\"] == annot_gene[\"gene_endpos\"])):\n",
    "                                del annot_gene[\"gene_startpos\"]\n",
    "                                del annot_gene[\"gene_endpos\"]\n",
    "                            elif (annot_gene[\"gene_startpos\"] > annot_gene[\"gene_endpos\"]):\n",
    "                                stpos = annot_gene[\"gene_startpos\"]\n",
    "                                annot_gene[\"gene_startpos\"] = annot_gene[\"gene_endpos\"]\n",
    "                                annot_gene[\"gene_endpos\"] = stpos\n",
    "                        else:\n",
    "                            del annot_gene[\"gene_startpos\"]\n",
    "                    elif \"gene_endpos\" in annot_gene:\n",
    "                        del annot_gene[\"gene_endpos\"]\n",
    "                    annot_genes.append(annot_gene)\n",
    "                    if \"gene_id\" in annot_gene and re.match(\"^ctg\\\\d+_orf\\\\d+$\", annot_gene[\"gene_id\"]):\n",
    "                        continue\n",
    "                    newannotation_count += 1\n",
    "                    gn = \"\"\n",
    "                    if \"gene_id\" in annot_gene:\n",
    "                        gn = \"{}/\".format(annot_gene[\"gene_id\"])\n",
    "                    else:\n",
    "                        gn = \"-/\"\n",
    "                    if \"gene_name\" in annot_gene:\n",
    "                        gn = \"{}{}\".format(gn, annot_gene[\"gene_name\"])\n",
    "                    else:\n",
    "                        gn = \"{}-\".format(gn)\n",
    "                    newannotated_genes.append(gn)\n",
    "\n",
    "            annotation_data[\"general_params\"][\"genes\"][\"gene\"] = annot_genes            \n",
    "        \n",
    "        if newannotation_count > 0:\n",
    "            with open(\"../../preprocessed/new_annotated_genes.tsv\", \"a\") as o:\n",
    "                o.write(\"{}\\t{}\\n\".format(annotation_data[\"general_params\"][\"mibig_accession\"], \";\".join(newannotated_genes)))\n",
    "        \n",
    "        with open(\"../../preprocessed/gbks_gene_counts.tsv\", \"a\") as o:\n",
    "            report_row = (\"{}\\t{}\\t{}-{}\\t{}\\t{}\\t{}\\t{}\".format(annotation_data[\"general_params\"][\"mibig_accession\"], record[\"nucl_acc\"], record[\"start_pos\"], record[\"end_pos\"], len(genes_in_gbk.keys()), len(genes_from_mibig.keys()), newannotation_count, reannotation_count))\n",
    "            o.write(\"{}\\n\".format(report_row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0000001\n",
      "Scanning BGC0000002\n",
      "Scanning BGC0000003\n",
      "Scanning BGC0000004\n",
      "Scanning BGC0000005\n",
      "Scanning BGC0000006\n",
      "Scanning BGC0000007\n",
      "Scanning BGC0000008\n",
      "Scanning BGC0000009\n",
      "Scanning BGC0000010\n",
      "Scanning BGC0000011\n",
      "Scanning BGC0000012\n",
      "Scanning BGC0000013\n",
      "Scanning BGC0000014\n",
      "Scanning BGC0000015\n",
      "Scanning BGC0000016\n",
      "Scanning BGC0000017\n",
      "Scanning BGC0000018\n",
      "Scanning BGC0000019\n",
      "Scanning BGC0000020\n",
      "Scanning BGC0000021\n",
      "Scanning BGC0000022\n",
      "Scanning BGC0000023\n",
      "Scanning BGC0000024\n",
      "Scanning BGC0000025\n",
      "Scanning BGC0000026\n",
      "Scanning BGC0000027\n",
      "Scanning BGC0000028\n",
      "Scanning BGC0000029\n",
      "Scanning BGC0000030\n",
      "Scanning BGC0000031\n",
      "Scanning BGC0000032\n",
      "Scanning BGC0000033\n",
      "Scanning BGC0000034\n",
      "Scanning BGC0000035\n",
      "Scanning BGC0000036\n",
      "Scanning BGC0000037\n",
      "Scanning BGC0000038\n",
      "Error...\n",
      "Scanning BGC0000039\n",
      "Scanning BGC0000040\n",
      "Scanning BGC0000041\n",
      "Scanning BGC0000042\n",
      "Scanning BGC0000043\n",
      "Scanning BGC0000044\n",
      "Scanning BGC0000045\n",
      "Scanning BGC0000046\n",
      "Scanning BGC0000047\n",
      "Scanning BGC0000048\n",
      "Scanning BGC0000049\n",
      "Scanning BGC0000050\n",
      "Scanning BGC0000051\n",
      "Scanning BGC0000052\n",
      "Scanning BGC0000053\n",
      "Scanning BGC0000054\n",
      "Scanning BGC0000055\n",
      "Scanning BGC0000056\n",
      "Scanning BGC0000057\n",
      "Scanning BGC0000058\n",
      "Scanning BGC0000059\n",
      "Scanning BGC0000060\n",
      "Scanning BGC0000061\n",
      "Scanning BGC0000062\n",
      "Scanning BGC0000063\n",
      "Scanning BGC0000064\n",
      "Scanning BGC0000065\n",
      "Scanning BGC0000066\n",
      "Scanning BGC0000067\n",
      "Scanning BGC0000068\n",
      "Scanning BGC0000069\n",
      "Scanning BGC0000070\n",
      "Scanning BGC0000072\n",
      "Scanning BGC0000073\n",
      "Scanning BGC0000074\n",
      "Scanning BGC0000075\n",
      "Scanning BGC0000076\n",
      "Scanning BGC0000077\n",
      "Scanning BGC0000078\n",
      "Scanning BGC0000079\n",
      "Scanning BGC0000080\n",
      "Scanning BGC0000081\n",
      "Scanning BGC0000082\n",
      "Scanning BGC0000083\n",
      "Scanning BGC0000084\n",
      "Scanning BGC0000085\n",
      "Scanning BGC0000086\n",
      "Scanning BGC0000087\n",
      "Scanning BGC0000088\n",
      "Scanning BGC0000089\n",
      "Scanning BGC0000090\n",
      "Scanning BGC0000091\n",
      "Scanning BGC0000092\n",
      "Scanning BGC0000093\n",
      "Scanning BGC0000094\n",
      "Scanning BGC0000095\n",
      "Scanning BGC0000096\n",
      "Scanning BGC0000097\n",
      "Scanning BGC0000098\n",
      "Scanning BGC0000099\n",
      "Scanning BGC0000100\n",
      "Scanning BGC0000101\n",
      "Scanning BGC0000102\n",
      "Scanning BGC0000103\n",
      "Scanning BGC0000104\n",
      "Scanning BGC0000105\n",
      "Scanning BGC0000106\n",
      "Scanning BGC0000107\n",
      "Scanning BGC0000108\n",
      "Scanning BGC0000109\n",
      "Scanning BGC0000110\n",
      "Scanning BGC0000111\n",
      "Scanning BGC0000112\n",
      "Scanning BGC0000113\n",
      "Scanning BGC0000114\n",
      "Scanning BGC0000115\n",
      "Scanning BGC0000116\n",
      "Scanning BGC0000117\n",
      "Scanning BGC0000118\n",
      "Scanning BGC0000119\n",
      "Scanning BGC0000120\n",
      "Scanning BGC0000121\n",
      "Scanning BGC0000122\n",
      "Scanning BGC0000123\n",
      "Scanning BGC0000124\n",
      "Scanning BGC0000125\n",
      "Scanning BGC0000126\n",
      "Scanning BGC0000127\n",
      "Scanning BGC0000128\n",
      "Scanning BGC0000129\n",
      "Scanning BGC0000130\n",
      "Scanning BGC0000131\n",
      "Scanning BGC0000132\n",
      "Scanning BGC0000133\n",
      "Scanning BGC0000134\n",
      "Scanning BGC0000135\n",
      "Scanning BGC0000136\n",
      "Scanning BGC0000137\n",
      "Scanning BGC0000138\n",
      "Scanning BGC0000140\n",
      "Scanning BGC0000141\n",
      "Scanning BGC0000142\n",
      "Scanning BGC0000143\n",
      "Scanning BGC0000144\n",
      "Scanning BGC0000145\n",
      "Scanning BGC0000146\n",
      "Scanning BGC0000147\n",
      "Scanning BGC0000148\n",
      "Scanning BGC0000149\n",
      "Scanning BGC0000150\n",
      "Scanning BGC0000151\n",
      "Scanning BGC0000152\n",
      "Scanning BGC0000153\n",
      "Scanning BGC0000154\n",
      "Scanning BGC0000155\n",
      "Scanning BGC0000156\n",
      "Scanning BGC0000157\n",
      "Scanning BGC0000158\n",
      "Scanning BGC0000159\n",
      "Scanning BGC0000160\n",
      "Scanning BGC0000161\n",
      "Scanning BGC0000162\n",
      "Scanning BGC0000163\n",
      "Scanning BGC0000164\n",
      "Scanning BGC0000165\n",
      "Scanning BGC0000166\n",
      "Scanning BGC0000167\n",
      "Scanning BGC0000168\n",
      "Scanning BGC0000170\n",
      "Scanning BGC0000171\n",
      "Scanning BGC0000172\n",
      "Scanning BGC0000173\n",
      "Scanning BGC0000174\n",
      "Scanning BGC0000175\n",
      "Scanning BGC0000176\n",
      "Scanning BGC0000177\n",
      "Scanning BGC0000178\n",
      "Scanning BGC0000179\n",
      "Scanning BGC0000180\n",
      "Scanning BGC0000181\n",
      "Scanning BGC0000182\n",
      "Scanning BGC0000183\n",
      "Scanning BGC0000184\n",
      "Scanning BGC0000185\n",
      "Scanning BGC0000186\n",
      "Scanning BGC0000187\n",
      "Scanning BGC0000188\n",
      "Scanning BGC0000189\n",
      "Scanning BGC0000190\n",
      "Scanning BGC0000191\n",
      "Scanning BGC0000192\n",
      "Scanning BGC0000193\n",
      "Scanning BGC0000194\n",
      "Scanning BGC0000195\n",
      "Scanning BGC0000196\n",
      "Scanning BGC0000197\n",
      "Scanning BGC0000198\n",
      "Scanning BGC0000199\n",
      "Scanning BGC0000200\n",
      "Scanning BGC0000201\n",
      "Scanning BGC0000202\n",
      "Scanning BGC0000203\n",
      "Scanning BGC0000204\n",
      "Scanning BGC0000205\n",
      "Scanning BGC0000206\n",
      "Scanning BGC0000207\n",
      "Scanning BGC0000208\n",
      "Scanning BGC0000209\n",
      "Scanning BGC0000210\n",
      "Scanning BGC0000211\n",
      "Scanning BGC0000212\n",
      "Scanning BGC0000213\n",
      "Scanning BGC0000214\n",
      "Scanning BGC0000215\n",
      "Scanning BGC0000216\n",
      "Scanning BGC0000217\n",
      "Scanning BGC0000218\n",
      "Scanning BGC0000219\n",
      "Scanning BGC0000220\n",
      "Scanning BGC0000221\n",
      "Scanning BGC0000222\n",
      "Scanning BGC0000223\n",
      "Scanning BGC0000224\n",
      "Scanning BGC0000225\n",
      "Scanning BGC0000226\n",
      "Scanning BGC0000227\n",
      "Scanning BGC0000228\n",
      "Scanning BGC0000229\n",
      "Scanning BGC0000230\n",
      "Scanning BGC0000231\n",
      "Scanning BGC0000232\n",
      "Scanning BGC0000233\n",
      "Scanning BGC0000234\n",
      "Scanning BGC0000235\n",
      "Scanning BGC0000236\n",
      "Scanning BGC0000237\n",
      "Scanning BGC0000238\n",
      "Scanning BGC0000239\n",
      "Scanning BGC0000240\n",
      "Scanning BGC0000241\n",
      "Scanning BGC0000242\n",
      "Scanning BGC0000243\n",
      "Scanning BGC0000244\n",
      "Scanning BGC0000245\n",
      "Scanning BGC0000246\n",
      "Scanning BGC0000247\n",
      "Scanning BGC0000248\n",
      "Scanning BGC0000249\n",
      "Scanning BGC0000250\n",
      "Scanning BGC0000251\n",
      "Scanning BGC0000252\n",
      "Scanning BGC0000253\n",
      "Scanning BGC0000254\n",
      "Scanning BGC0000255\n",
      "Scanning BGC0000256\n",
      "Scanning BGC0000257\n",
      "Scanning BGC0000258\n",
      "Scanning BGC0000259\n",
      "Scanning BGC0000260\n",
      "Scanning BGC0000261\n",
      "Scanning BGC0000262\n",
      "Scanning BGC0000263\n",
      "Scanning BGC0000264\n",
      "Scanning BGC0000265\n",
      "Scanning BGC0000266\n",
      "Scanning BGC0000267\n",
      "Scanning BGC0000268\n",
      "Scanning BGC0000269\n",
      "Scanning BGC0000270\n",
      "Scanning BGC0000271\n",
      "Scanning BGC0000272\n",
      "Scanning BGC0000273\n",
      "Scanning BGC0000274\n",
      "Scanning BGC0000275\n",
      "Scanning BGC0000276\n",
      "Scanning BGC0000277\n",
      "Scanning BGC0000278\n",
      "Scanning BGC0000279\n",
      "Scanning BGC0000280\n",
      "Scanning BGC0000281\n",
      "Scanning BGC0000282\n",
      "Scanning BGC0000283\n",
      "Scanning BGC0000284\n",
      "Scanning BGC0000285\n",
      "Scanning BGC0000286\n",
      "Scanning BGC0000287\n",
      "Scanning BGC0000288\n",
      "Scanning BGC0000289\n",
      "Scanning BGC0000290\n",
      "Scanning BGC0000291\n",
      "Scanning BGC0000292\n",
      "Scanning BGC0000293\n",
      "Scanning BGC0000294\n",
      "Scanning BGC0000295\n",
      "Scanning BGC0000296\n",
      "Scanning BGC0000297\n",
      "Scanning BGC0000298\n",
      "Scanning BGC0000299\n",
      "Scanning BGC0000300\n",
      "Scanning BGC0000301\n",
      "Scanning BGC0000302\n",
      "Scanning BGC0000303\n",
      "Scanning BGC0000304\n",
      "Scanning BGC0000305\n",
      "Scanning BGC0000306\n",
      "Scanning BGC0000307\n",
      "Scanning BGC0000308\n",
      "Scanning BGC0000309\n",
      "Scanning BGC0000310\n",
      "Scanning BGC0000311\n",
      "Scanning BGC0000312\n",
      "Scanning BGC0000313\n",
      "Scanning BGC0000314\n",
      "Scanning BGC0000315\n",
      "Scanning BGC0000316\n",
      "Scanning BGC0000317\n",
      "Scanning BGC0000318\n",
      "Scanning BGC0000319\n",
      "Scanning BGC0000320\n",
      "Scanning BGC0000321\n",
      "Scanning BGC0000322\n",
      "Scanning BGC0000323\n",
      "Scanning BGC0000324\n",
      "Scanning BGC0000325\n",
      "Scanning BGC0000326\n",
      "Scanning BGC0000327\n",
      "Scanning BGC0000328\n",
      "Scanning BGC0000329\n",
      "Scanning BGC0000330\n",
      "Scanning BGC0000331\n",
      "Scanning BGC0000332\n",
      "Scanning BGC0000333\n",
      "Scanning BGC0000334\n",
      "Scanning BGC0000335\n",
      "Scanning BGC0000336\n",
      "Scanning BGC0000337\n",
      "Scanning BGC0000338\n",
      "Scanning BGC0000339\n",
      "Scanning BGC0000340\n",
      "Scanning BGC0000341\n",
      "Scanning BGC0000342\n",
      "Scanning BGC0000343\n",
      "Scanning BGC0000344\n",
      "Scanning BGC0000345\n",
      "Scanning BGC0000346\n",
      "Scanning BGC0000347\n",
      "Scanning BGC0000348\n",
      "Scanning BGC0000349\n",
      "Scanning BGC0000350\n",
      "Scanning BGC0000351\n",
      "Scanning BGC0000352\n",
      "Scanning BGC0000353\n",
      "Scanning BGC0000354\n",
      "Scanning BGC0000355\n",
      "Scanning BGC0000356\n",
      "Scanning BGC0000357\n",
      "Scanning BGC0000358\n",
      "Scanning BGC0000359\n",
      "Scanning BGC0000360\n",
      "Scanning BGC0000361\n",
      "Scanning BGC0000362\n",
      "Scanning BGC0000363\n",
      "Scanning BGC0000364\n",
      "Scanning BGC0000365\n",
      "Scanning BGC0000366\n",
      "Scanning BGC0000367\n",
      "Scanning BGC0000368\n",
      "Scanning BGC0000369\n",
      "Scanning BGC0000370\n",
      "Scanning BGC0000371\n",
      "Scanning BGC0000372\n",
      "Scanning BGC0000373\n",
      "Scanning BGC0000374\n",
      "Scanning BGC0000375\n",
      "Scanning BGC0000376\n",
      "Scanning BGC0000377\n",
      "Scanning BGC0000378\n",
      "Scanning BGC0000379\n",
      "Scanning BGC0000380\n",
      "Scanning BGC0000381\n",
      "Scanning BGC0000382\n",
      "Scanning BGC0000383\n",
      "Scanning BGC0000384\n",
      "Scanning BGC0000385\n",
      "Scanning BGC0000386\n",
      "Scanning BGC0000387\n",
      "Scanning BGC0000388\n",
      "Scanning BGC0000389\n",
      "Scanning BGC0000391\n",
      "Scanning BGC0000392\n",
      "Scanning BGC0000393\n",
      "Scanning BGC0000394\n",
      "Scanning BGC0000395\n",
      "Scanning BGC0000396\n",
      "Scanning BGC0000397\n",
      "Scanning BGC0000398\n",
      "Scanning BGC0000399\n",
      "Scanning BGC0000400\n",
      "Scanning BGC0000401\n",
      "Scanning BGC0000402\n",
      "Scanning BGC0000403\n",
      "Scanning BGC0000404\n",
      "Scanning BGC0000405\n",
      "Scanning BGC0000406\n",
      "Scanning BGC0000407\n",
      "Scanning BGC0000408\n",
      "Scanning BGC0000409\n",
      "Scanning BGC0000410\n",
      "Scanning BGC0000411\n",
      "Scanning BGC0000412\n",
      "Scanning BGC0000413\n",
      "Scanning BGC0000414\n",
      "Scanning BGC0000415\n",
      "Scanning BGC0000416\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0000417\n",
      "Scanning BGC0000418\n",
      "Scanning BGC0000419\n",
      "Scanning BGC0000420\n",
      "Scanning BGC0000421\n",
      "Scanning BGC0000422\n",
      "Scanning BGC0000423\n",
      "Scanning BGC0000424\n",
      "Scanning BGC0000425\n",
      "Scanning BGC0000426\n",
      "Scanning BGC0000427\n",
      "Scanning BGC0000428\n",
      "Scanning BGC0000429\n",
      "Scanning BGC0000430\n",
      "Scanning BGC0000431\n",
      "Scanning BGC0000432\n",
      "Scanning BGC0000433\n",
      "Scanning BGC0000434\n",
      "Scanning BGC0000435\n",
      "Scanning BGC0000436\n",
      "Scanning BGC0000437\n",
      "Scanning BGC0000438\n",
      "Scanning BGC0000439\n",
      "Scanning BGC0000440\n",
      "Scanning BGC0000441\n",
      "Scanning BGC0000442\n",
      "Scanning BGC0000443\n",
      "Scanning BGC0000444\n",
      "Scanning BGC0000445\n",
      "Scanning BGC0000446\n",
      "Scanning BGC0000447\n",
      "Scanning BGC0000448\n",
      "Scanning BGC0000449\n",
      "Scanning BGC0000450\n",
      "Scanning BGC0000451\n",
      "Scanning BGC0000452\n",
      "Scanning BGC0000453\n",
      "Scanning BGC0000454\n",
      "Scanning BGC0000455\n",
      "Scanning BGC0000456\n",
      "Scanning BGC0000457\n",
      "Scanning BGC0000458\n",
      "Scanning BGC0000459\n",
      "Scanning BGC0000460\n",
      "Scanning BGC0000461\n",
      "Scanning BGC0000462\n",
      "Scanning BGC0000463\n",
      "Scanning BGC0000464\n",
      "Scanning BGC0000465\n",
      "Scanning BGC0000466\n",
      "Scanning BGC0000467\n",
      "Scanning BGC0000468\n",
      "Scanning BGC0000469\n",
      "Scanning BGC0000470\n",
      "Scanning BGC0000471\n",
      "Scanning BGC0000472\n",
      "Scanning BGC0000473\n",
      "Scanning BGC0000474\n",
      "Scanning BGC0000475\n",
      "Scanning BGC0000476\n",
      "Scanning BGC0000477\n",
      "Scanning BGC0000478\n",
      "Scanning BGC0000479\n",
      "Scanning BGC0000480\n",
      "Scanning BGC0000481\n",
      "Scanning BGC0000482\n",
      "Scanning BGC0000483\n",
      "Scanning BGC0000484\n",
      "Scanning BGC0000485\n",
      "Scanning BGC0000486\n",
      "Scanning BGC0000487\n",
      "Scanning BGC0000488\n",
      "Scanning BGC0000489\n",
      "Scanning BGC0000490\n",
      "Scanning BGC0000491\n",
      "Scanning BGC0000492\n",
      "Scanning BGC0000493\n",
      "Scanning BGC0000494\n",
      "Scanning BGC0000495\n",
      "Scanning BGC0000496\n",
      "Scanning BGC0000497\n",
      "Scanning BGC0000498\n",
      "Scanning BGC0000499\n",
      "Scanning BGC0000500\n",
      "Scanning BGC0000501\n",
      "Scanning BGC0000502\n",
      "Scanning BGC0000503\n",
      "Scanning BGC0000504\n",
      "Scanning BGC0000505\n",
      "Scanning BGC0000506\n",
      "Scanning BGC0000507\n",
      "Scanning BGC0000508\n",
      "Scanning BGC0000509\n",
      "Scanning BGC0000510\n",
      "Scanning BGC0000511\n",
      "Scanning BGC0000513\n",
      "Scanning BGC0000514\n",
      "Scanning BGC0000515\n",
      "Scanning BGC0000516\n",
      "Scanning BGC0000517\n",
      "Scanning BGC0000518\n",
      "Scanning BGC0000519\n",
      "Scanning BGC0000520\n",
      "Scanning BGC0000521\n",
      "Scanning BGC0000522\n",
      "Scanning BGC0000523\n",
      "Scanning BGC0000525\n",
      "Scanning BGC0000526\n",
      "Scanning BGC0000527\n",
      "Scanning BGC0000528\n",
      "Scanning BGC0000529\n",
      "Scanning BGC0000530\n",
      "Scanning BGC0000531\n",
      "Scanning BGC0000532\n",
      "Scanning BGC0000533\n",
      "Scanning BGC0000534\n",
      "Scanning BGC0000535\n",
      "Scanning BGC0000536\n",
      "Scanning BGC0000537\n",
      "Scanning BGC0000538\n",
      "Scanning BGC0000539\n",
      "Scanning BGC0000540\n",
      "Scanning BGC0000541\n",
      "Scanning BGC0000542\n",
      "Scanning BGC0000543\n",
      "Scanning BGC0000544\n",
      "Scanning BGC0000545\n",
      "Scanning BGC0000546\n",
      "Scanning BGC0000547\n",
      "Scanning BGC0000548\n",
      "Scanning BGC0000549\n",
      "Scanning BGC0000550\n",
      "Scanning BGC0000551\n",
      "Scanning BGC0000552\n",
      "Scanning BGC0000553\n",
      "Scanning BGC0000554\n",
      "Scanning BGC0000555\n",
      "Scanning BGC0000556\n",
      "Scanning BGC0000557\n",
      "Scanning BGC0000558\n",
      "Scanning BGC0000559\n",
      "Scanning BGC0000560\n",
      "Scanning BGC0000561\n",
      "Scanning BGC0000562\n",
      "Scanning BGC0000563\n",
      "Scanning BGC0000564\n",
      "Scanning BGC0000565\n",
      "Scanning BGC0000566\n",
      "Scanning BGC0000567\n",
      "Scanning BGC0000568\n",
      "Scanning BGC0000569\n",
      "Scanning BGC0000570\n",
      "Scanning BGC0000571\n",
      "Scanning BGC0000572\n",
      "Scanning BGC0000573\n",
      "Scanning BGC0000574\n",
      "Scanning BGC0000575\n",
      "Scanning BGC0000576\n",
      "Scanning BGC0000577\n",
      "Scanning BGC0000578\n",
      "Scanning BGC0000579\n",
      "Scanning BGC0000580\n",
      "Scanning BGC0000581\n",
      "Scanning BGC0000582\n",
      "Scanning BGC0000583\n",
      "Scanning BGC0000584\n",
      "Scanning BGC0000585\n",
      "Scanning BGC0000586\n",
      "Scanning BGC0000587\n",
      "Scanning BGC0000588\n",
      "Scanning BGC0000589\n",
      "Scanning BGC0000590\n",
      "Scanning BGC0000591\n",
      "Scanning BGC0000592\n",
      "Scanning BGC0000593\n",
      "Scanning BGC0000594\n",
      "Scanning BGC0000595\n",
      "Scanning BGC0000596\n",
      "Scanning BGC0000597\n",
      "Scanning BGC0000598\n",
      "Scanning BGC0000599\n",
      "Scanning BGC0000600\n",
      "Scanning BGC0000601\n",
      "Scanning BGC0000602\n",
      "Scanning BGC0000603\n",
      "Scanning BGC0000604\n",
      "Scanning BGC0000605\n",
      "Scanning BGC0000606\n",
      "Scanning BGC0000607\n",
      "Scanning BGC0000608\n",
      "Scanning BGC0000609\n",
      "Scanning BGC0000610\n",
      "Scanning BGC0000611\n",
      "Scanning BGC0000612\n",
      "Scanning BGC0000613\n",
      "Scanning BGC0000614\n",
      "Scanning BGC0000615\n",
      "Scanning BGC0000616\n",
      "Scanning BGC0000617\n",
      "Scanning BGC0000618\n",
      "Scanning BGC0000619\n",
      "Scanning BGC0000620\n",
      "Scanning BGC0000621\n",
      "Scanning BGC0000622\n",
      "Scanning BGC0000623\n",
      "Scanning BGC0000624\n",
      "Scanning BGC0000625\n",
      "Scanning BGC0000626\n",
      "Scanning BGC0000627\n",
      "Scanning BGC0000628\n",
      "Scanning BGC0000629\n",
      "Scanning BGC0000630\n",
      "Scanning BGC0000631\n",
      "Scanning BGC0000632\n",
      "Scanning BGC0000633\n",
      "Scanning BGC0000634\n",
      "Scanning BGC0000635\n",
      "Scanning BGC0000636\n",
      "Scanning BGC0000637\n",
      "Scanning BGC0000638\n",
      "Scanning BGC0000639\n",
      "Scanning BGC0000640\n",
      "Scanning BGC0000641\n",
      "Scanning BGC0000642\n",
      "Scanning BGC0000643\n",
      "Scanning BGC0000644\n",
      "Scanning BGC0000645\n",
      "Scanning BGC0000646\n",
      "Scanning BGC0000647\n",
      "Scanning BGC0000648\n",
      "Scanning BGC0000649\n",
      "Scanning BGC0000650\n",
      "Scanning BGC0000651\n",
      "Scanning BGC0000652\n",
      "Scanning BGC0000653\n",
      "Scanning BGC0000654\n",
      "Scanning BGC0000655\n",
      "Scanning BGC0000656\n",
      "Scanning BGC0000657\n",
      "Scanning BGC0000658\n",
      "Scanning BGC0000659\n",
      "Scanning BGC0000660\n",
      "Scanning BGC0000661\n",
      "Scanning BGC0000662\n",
      "Scanning BGC0000663\n",
      "Scanning BGC0000664\n",
      "Scanning BGC0000665\n",
      "Scanning BGC0000666\n",
      "Scanning BGC0000667\n",
      "Scanning BGC0000668\n",
      "Scanning BGC0000669\n",
      "Scanning BGC0000670\n",
      "Scanning BGC0000671\n",
      "Scanning BGC0000672\n",
      "Scanning BGC0000673\n",
      "Scanning BGC0000674\n",
      "Scanning BGC0000675\n",
      "Scanning BGC0000676\n",
      "Scanning BGC0000677\n",
      "Scanning BGC0000678\n",
      "Scanning BGC0000679\n",
      "Scanning BGC0000680\n",
      "Scanning BGC0000682\n",
      "Scanning BGC0000683\n"
     ]
    }
   ],
   "source": [
    "input_path = \"../../preprocessed/json_2.0_phase_5/\"\n",
    "output_folder = \"../../preprocessed/json_2.0_genefinding/\"\n",
    "for json_path in sorted(glob.glob(path.join(input_path, \"BGC*.json\"))):\n",
    "    with open(json_path, \"r\") as json_file:\n",
    "        bgc_id = path.basename(json_path).split(\".\")[0]\n",
    "        data = json.load(json_file)\n",
    "        print(\"Scanning {}\".format(bgc_id))\n",
    "        #fetch_additional_genes(data)\n",
    "        with open(path.join(output_folder, \"{}.json\".format(bgc_id)), \"w\") as o:\n",
    "            o.write(json.dumps(data, indent=4, separators=(',', ': ')))\n",
    "\n",
    "print(\"All data fetched!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_counts = {}\n",
    "with open(\"../../preprocessed/gbks_gene_counts.tsv\", \"r\") as o:\n",
    "    i = 0\n",
    "    for line in o:\n",
    "        if i == 0:\n",
    "            continue\n",
    "        i += 1\n",
    "        line = line.rstrip()\n",
    "        tabs = line.split(\"\\t\")\n",
    "        bgcid = tabs[0]\n",
    "        if bgcid not in gene_counts:\n",
    "            gene_counts[bgcid] = [0, 0, 0, 0] # total loci, total genes from acc, total genes from mibig, total genes\n",
    "        gene_counts[bgcid][0] += 1\n",
    "        gene_counts[bgcid][1] += int(tabs[3])\n",
    "        gene_counts[bgcid][2] += int(tabs[4])\n",
    "        gene_counts[bgcid][3] += int(tabs[3]) + int(tabs[4])\n",
    "        \n",
    "with open(\"../../preprocessed/gbks_gene_counts_merged.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\ttotal_loci\\tfrom_acc\\tfrom_mibig\\ttotal\\n\")\n",
    "    for bgcid in sorted(gene_counts.keys()):\n",
    "        ar = gene_counts[bgcid]\n",
    "        o.write(\"{}\\t{}\\t{}\\t{}\\t{}\\n\".format(bgcid, ar[0], ar[1], ar[2], ar[3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
