{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "###### phase 7: re-check gene annotations, comparing it to the GBK, filter out ones without any new information added ######\n",
    "#### it also populate organism and taxon properties from mibig finalgbk ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import path, makedirs\n",
    "import glob\n",
    "import json\n",
    "from jsonschema import validate, Draft7Validator\n",
    "from tempfile import TemporaryDirectory\n",
    "from Bio import SeqIO, Entrez\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Alphabet import generic_protein\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import sys\n",
    "import datetime\n",
    "import time\n",
    "import re\n",
    "import urllib3\n",
    "from urllib.parse import quote\n",
    "import certifi\n",
    "from xml.dom import minidom\n",
    "http = urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=certifi.where())\n",
    "from jsonschema.validators import Draft7Validator\n",
    "from shutil import copy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ncbi_tax_ids(organism_name, email):\n",
    "    tax_ids = []\n",
    "    \n",
    "    Entrez.email = email\n",
    "    num_try = 1\n",
    "    while num_try < 6:\n",
    "        try:\n",
    "            organism_name_escaped = \"\\\"{}\\\"\".format(organism_name)\n",
    "            dom = minidom.parse(Entrez.esearch(db=\"taxonomy\", term=organism_name_escaped))\n",
    "            ids = dom.getElementsByTagName('Id')\n",
    "            if len(ids) < 1:\n",
    "                raise Exception()\n",
    "            for tax_id in ids:\n",
    "                tax_ids.append(tax_id.firstChild.nodeValue)\n",
    "            break\n",
    "        except:\n",
    "            pass\n",
    "        num_try += 1\n",
    "        time.sleep(5)\n",
    "    \n",
    "    return tax_ids\n",
    "\n",
    "def fetch_gbk(nucl_acc, email, clean_cache = False):\n",
    "    cache_folder = \"../../preprocessed/cache/cached_gbks/\"\n",
    "    cache_path = path.join(cache_folder, \"{}.gbk\".format(nucl_acc))\n",
    "\n",
    "    if not path.exists(cache_folder):\n",
    "        makedirs(cache_folder)\n",
    "    \n",
    "    if clean_cache or not path.exists(cache_path) or path.getsize(cache_path) < 100:\n",
    "        if nucl_acc.startswith(\"MIBIG.BGC\"):\n",
    "            [_, bgc_id, cl_id] = nucl_acc.split(\".\")\n",
    "            mibig_final_file_path = \"../../inputs/cached_mibig_finalgbks/{}.1.final.gbk\".format(bgc_id)\n",
    "            with open(mibig_final_file_path, \"r\") as mibig_final_file:\n",
    "                with open(cache_path, \"w\", encoding=\"utf-8\") as cache_gbk_file:\n",
    "                    write_to_file = False\n",
    "                    for line in mibig_final_file:\n",
    "                        if line.startswith(\"LOCUS\"):\n",
    "                            gbk_accession = line.split(\" \")[1]\n",
    "                            if gbk_accession == \"{}.{}\".format(bgc_id, cl_id):\n",
    "                                write_to_file = True\n",
    "                        if write_to_file:\n",
    "                            cache_gbk_file.write(line)\n",
    "                        if line.rstrip() == \"//\":\n",
    "                            break\n",
    "        else: # (re)download from ncbi\n",
    "            Entrez.email = email\n",
    "            handle = Entrez.efetch(db=\"nucleotide\", id=nucl_acc, rettype=\"gbwithparts\", retmode=\"text\")\n",
    "            if not path.exists(cache_folder):\n",
    "                makedirs(cache_folder)\n",
    "            with open(cache_path, \"w\") as gbk_file:\n",
    "                gbk_file.write(handle.read())\n",
    "            \n",
    "    return open(cache_path, \"r\")\n",
    "\n",
    "# from antismash\n",
    "def get_aa_translation(seq_record, feature):\n",
    "    \"\"\"Obtain content for translation qualifier for specific CDS feature in sequence record\"\"\"\n",
    "    extracted = feature.extract(seq_record.seq).ungap('-')\n",
    "\n",
    "    # ensure the extracted section is a multiple of three by trimming any excess\n",
    "    if len(extracted) % 3 != 0:\n",
    "        extracted = extracted[:-(len(extracted) % 3)]\n",
    "\n",
    "    fasta_seq = extracted.translate(to_stop=True)\n",
    "    if len(fasta_seq) == 0:\n",
    "        print(\"Retranslating {} with stop codons\".format(feature.id))\n",
    "        fasta_seq = extracted.translate()\n",
    "\n",
    "    # replace ambiguous aminos with an explicit unknown\n",
    "    string_version = str(fasta_seq)\n",
    "    for bad in \"*BJOUZ\":\n",
    "        string_version = string_version.replace(bad, \"X\")\n",
    "\n",
    "    # and remove any gaps\n",
    "    string_version = string_version.replace(\"-\", \"\")\n",
    "    fasta_seq = Seq(string_version, generic_protein)\n",
    "\n",
    "    return fasta_seq\n",
    "\n",
    "def get_aa_sequence(feature, to_stop=False):\n",
    "    \"\"\"Extract sequence from specific CDS feature in sequence record\"\"\"\n",
    "    fasta_seq = feature.qualifiers['translation'][0]\n",
    "    if \"*\" in fasta_seq:\n",
    "        if to_stop:\n",
    "            fasta_seq = fasta_seq.split('*')[0]\n",
    "        else:\n",
    "            fasta_seq = fasta_seq.replace(\"*\",\"X\")\n",
    "    if \"-\" in fasta_seq:\n",
    "        fasta_seq = fasta_seq.replace(\"-\",\"\")\n",
    "    return fasta_seq\n",
    "\n",
    "\n",
    "def count_props(input_dict, cur_path, result):\n",
    "    \"\"\"given a (mibig?) json, construct a list of property paths\n",
    "    along with its presence count in the json object\"\"\"\n",
    "    key_path = cur_path\n",
    "    \n",
    "    if isinstance(input_dict, dict):\n",
    "        for key in input_dict.keys():\n",
    "            result = count_props(input_dict[key], \"{}/{}\".format(key_path, key), result)\n",
    "    elif isinstance(input_dict, list):\n",
    "        key_path = \"{}[]\".format(key_path)\n",
    "        for node in input_dict:\n",
    "            result = count_props(node, \"{}\".format(key_path), result)\n",
    "\n",
    "    if not isinstance(input_dict, dict):\n",
    "        if key_path not in result:\n",
    "            result[key_path] = 0\n",
    "        result[key_path] += 1\n",
    "    \n",
    "    return result\n",
    "\n",
    "\n",
    "def fetch_props_new_schema(input_dict, cur_path, result):\n",
    "    \"\"\"given a (mibig?) json draft7 schema, construct a list of property paths\n",
    "    along with either required == True for each properties\"\"\"\n",
    "    key_path = cur_path\n",
    "    if (\"type\" not in input_dict) or (input_dict[\"type\"] not in [\"object\", \"array\"]):\n",
    "        key_path = \"{}\".format(cur_path) # string / etc.\n",
    "    elif input_dict[\"type\"] == \"object\":\n",
    "        for key in input_dict[\"properties\"]:\n",
    "            result = fetch_props_new_schema(input_dict[\"properties\"][key], \"{}/{}\".format(key_path, key), result)\n",
    "    elif input_dict[\"type\"] == \"array\":\n",
    "        key_path = \"{}[]\".format(cur_path)\n",
    "        result = fetch_props_new_schema(input_dict[\"items\"], \"{}\".format(key_path), result)\n",
    "    \n",
    "    if key_path not in result and \"properties\" not in input_dict:\n",
    "        result[key_path] = False # can't really use this\n",
    "    return result\n",
    "\n",
    "def sanitise_gene_name(name):\n",
    "    if name is None:\n",
    "        return None\n",
    "    name = str(name)\n",
    "    illegal_chars = set(\"!\\\"#$%&()*+,:; \\r\\n\\t=>?@[]^`'{|}/ \")\n",
    "    for char in set(name).intersection(illegal_chars):\n",
    "        name = name.replace(char, \"_\")\n",
    "    return name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_mibig_final_gbk(bgc_id, clean_cache = False):\n",
    "    return open(\"../../inputs/cached_mibig_finalgbks/{}.1.final.gbk\".format(bgc_id), \"r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _fetch_val_to_ref(input_dict, parent_dict, key, all_references):\n",
    "    if isinstance(input_dict, dict):\n",
    "        for key in input_dict.keys():\n",
    "            _fetch_val_to_ref(input_dict[key], input_dict, key, all_references)\n",
    "    elif isinstance(input_dict, list):\n",
    "        for i, node in enumerate(input_dict):\n",
    "            _fetch_val_to_ref(node, input_dict, i, all_references)\n",
    "    elif isinstance(input_dict, str):\n",
    "        if parent_dict != None:\n",
    "            valu = input_dict.upper()\n",
    "            if valu not in all_references:\n",
    "                all_references[valu] = []\n",
    "            all_references[valu].append((parent_dict, key))\n",
    "    return all_references\n",
    "\n",
    "with open(\"../../preprocessed/reports/p7-updated_cases.tsv\", \"w\") as uc:\n",
    "    uc.write(\"\")\n",
    "    \n",
    "def _make_gene_id_match_reference(input_dict, reference_ids, bgc_id):\n",
    "    all_references = _fetch_val_to_ref(input_dict, None, None, {})\n",
    "    all_rids = {}\n",
    "    for rid in reference_ids:\n",
    "        if len(rid) > 0:\n",
    "            all_rids[rid.upper()] = rid\n",
    "    matches = set(all_references.keys()).intersection(set(all_rids.keys()))    \n",
    "    for key in matches:\n",
    "        for ar in all_references[key]:\n",
    "            if ar[0][ar[1]] != all_rids[key]:\n",
    "                with open(\"../../preprocessed/reports/p7-updated_cases.tsv\", \"a\") as uc:\n",
    "                    uc.write(\"{}\\t{}\\t{}\\n\".format(bgc_id, ar[0][ar[1]], all_rids[key]))\n",
    "                ar[0][ar[1]] = all_rids[key]\n",
    "\n",
    "def make_gene_id_match_reference(data, reference_ids):\n",
    "    bgc_id = data[\"cluster\"][\"mibig_accession\"]\n",
    "    _make_gene_id_match_reference(data[\"cluster\"][\"compounds\"], reference_ids, bgc_id)\n",
    "    _make_gene_id_match_reference(data[\"cluster\"].get(\"genes\", {}), reference_ids, bgc_id)\n",
    "    _make_gene_id_match_reference(data[\"cluster\"].get(\"polyketide\", {}), reference_ids, bgc_id)\n",
    "    _make_gene_id_match_reference(data[\"cluster\"].get(\"nrp\", {}), reference_ids, bgc_id)\n",
    "    _make_gene_id_match_reference(data[\"cluster\"].get(\"ripp\", {}), reference_ids, bgc_id)\n",
    "    _make_gene_id_match_reference(data[\"cluster\"].get(\"terpene\", {}), reference_ids, bgc_id)\n",
    "    _make_gene_id_match_reference(data[\"cluster\"].get(\"saccharide\", {}), reference_ids, bgc_id)\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "removed_extra_genes = {}\n",
    "removed_annotations = {}\n",
    "removed_operons = {}\n",
    "no_taxid = {}\n",
    "added_mibig_genes = {}\n",
    "no_gbk = {}\n",
    "fixed = {}\n",
    "ripps_to_fix = {}\n",
    "\n",
    "nt_length_and_cds_count = {}\n",
    "\n",
    "def check_gene_annotations(data):\n",
    "    loci = data[\"cluster\"][\"loci\"]\n",
    "    annots = data[\"cluster\"].get(\"genes\", {})\n",
    "    gbk_acc = loci[\"accession\"].upper()\n",
    "    bgc_id = data[\"cluster\"][\"mibig_accession\"]\n",
    "    email = \"mibig@secondarymetabolites.org\"\n",
    "    gbk_record = None\n",
    "    if True:\n",
    "        num_try = 1\n",
    "        clean_cache = False\n",
    "        while num_try < 6:\n",
    "            try:\n",
    "                with fetch_gbk(gbk_acc, email, clean_cache) as gbk_handle:\n",
    "                    seq_record = SeqIO.read(gbk_handle, \"genbank\") # the gbk should contains only 1 file\n",
    "                    if len(seq_record.seq) < 1:\n",
    "                        raise Exception(\"Empty sequence record {}\".format(gbk_acc))\n",
    "                    gbk_record = seq_record\n",
    "                    break\n",
    "            except:\n",
    "                print(\"Error...\")\n",
    "                clean_cache = True\n",
    "            num_try += 1\n",
    "            time.sleep(5)\n",
    "        if not isinstance(gbk_record, SeqRecord): # failed to download NCBI data\n",
    "            print(\"{} Failed to download: {}\".format(data[\"cluster\"][\"mibig_accession\"], gbk_acc))\n",
    "            no_gbk[bgc_id] = gbk_acc\n",
    "            return {}\n",
    "\n",
    "    # fill up taxid\n",
    "    tax_ids = get_ncbi_tax_ids(gbk_record.annotations[\"organism\"], email)\n",
    "    if len(tax_ids) != 1:\n",
    "        no_taxid[bgc_id] = gbk_record.annotations[\"organism\"]\n",
    "        print(tax_ids)\n",
    "    else:\n",
    "        data[\"cluster\"][\"ncbi_tax_id\"] = tax_ids[0]\n",
    "        data[\"cluster\"][\"organism_name\"] = gbk_record.annotations[\"organism\"]\n",
    "         \n",
    "    # fetch all CDS inside the cluster\n",
    "    cluster_cds = {}\n",
    "    cluster_cds_ids = set([\"\", \"No protein ID\"]) # quick lookup for existing IDs\n",
    "    for feature in seq_record.features:\n",
    "        if feature.type == \"CDS\":\n",
    "            if (\"start_coord\" not in loci) or (feature.location.start >= loci[\"start_coord\"]-1 and feature.location.end <= loci[\"end_coord\"]):\n",
    "                if \"gene\" in feature.qualifiers:\n",
    "                    cluster_cds_ids.add(feature.qualifiers[\"gene\"][0])\n",
    "                    cluster_cds[feature.qualifiers[\"gene\"][0]] = feature\n",
    "                if \"protein_id\" in feature.qualifiers:\n",
    "                    cluster_cds_ids.add(feature.qualifiers[\"protein_id\"][0])\n",
    "                    cluster_cds[feature.qualifiers[\"protein_id\"][0]] = feature\n",
    "                if \"locus_tag\" in feature.qualifiers:\n",
    "                    cluster_cds_ids.add(feature.qualifiers[\"locus_tag\"][0])\n",
    "                    cluster_cds[feature.qualifiers[\"locus_tag\"][0]] = feature\n",
    "                \n",
    "    # fetch all CDS in MIBiG finalgbk\n",
    "    mibig_extra_genes = []\n",
    "    with fetch_mibig_final_gbk(bgc_id) as mibig_gbk_handle:\n",
    "        mibig_seq_records = SeqIO.parse(mibig_gbk_handle, \"genbank\")\n",
    "        for mibig_seq_record in mibig_seq_records:\n",
    "            # look for gene information\n",
    "            match = re.search(\"(between(?P<start>\\d+)-(?P<end>\\d+)ntfrom){0,1}GenBankID(?P<acc>[A-Z0-9_\\.]+)\\.\", mibig_seq_record.annotations['comment'].replace(\" \", \"\").replace(\"\\n\", \"\"))\n",
    "            if match:\n",
    "                gbk_acc_mibig = match.group(\"acc\").upper()\n",
    "                if gbk_acc_mibig == gbk_acc:\n",
    "                    offset = 0\n",
    "                    if match.group(\"start\"):\n",
    "                        offset = int(match.group(\"start\")) - 1\n",
    "                    for feature in mibig_seq_record.features:\n",
    "                        if feature.type == \"CDS\":\n",
    "                            ids = []\n",
    "                            if \"gene\" in feature.qualifiers:\n",
    "                                if feature.qualifiers[\"gene\"][0].upper() in [sanitise_gene_name(gid.upper()) for gid in cluster_cds_ids]:\n",
    "                                    continue\n",
    "                                ids.append(feature.qualifiers[\"gene\"][0])\n",
    "                                cluster_cds[feature.qualifiers[\"gene\"][0]] = feature\n",
    "                            if \"protein_id\" in feature.qualifiers:\n",
    "                                if feature.qualifiers[\"protein_id\"][0].upper() in [sanitise_gene_name(gid.upper()) for gid in cluster_cds_ids]:\n",
    "                                    continue\n",
    "                                ids.append(feature.qualifiers[\"protein_id\"][0])\n",
    "                                cluster_cds[feature.qualifiers[\"protein_id\"][0]] = feature\n",
    "                            if \"locus_tag\" in feature.qualifiers:\n",
    "                                if feature.qualifiers[\"locus_tag\"][0].upper() in [sanitise_gene_name(gid.upper()) for gid in cluster_cds_ids]:\n",
    "                                    continue\n",
    "                                ids.append(feature.qualifiers[\"locus_tag\"][0])\n",
    "                                cluster_cds[feature.qualifiers[\"locus_tag\"][0]] = feature\n",
    "                            cluster_cds_ids.update(set(ids))\n",
    "                            mibig_extra_genes.append(feature._shift(offset)) \n",
    "                            if bgc_id not in added_mibig_genes:\n",
    "                                added_mibig_genes[bgc_id] = []\n",
    "                            added_mibig_genes[bgc_id].append(\"/\".join([gid for gid in ids]))\n",
    "                            \n",
    "    # add mibig extra genes\n",
    "    for feature in mibig_extra_genes:\n",
    "        if \"extra_genes\" not in annots:\n",
    "            annots[\"extra_genes\"] = []\n",
    "        if \"annotations\" not in annots:\n",
    "            annots[\"annotations\"] = []\n",
    "        gene_id = feature.qualifiers.get(\"locus_tag\", feature.qualifiers.get(\"protein_id\", feature.qualifiers.get(\"gene\", [None])))[0]\n",
    "        if gene_id != None:\n",
    "            extra_gene = {\n",
    "                \"id\": gene_id,\n",
    "                \"location\": {\n",
    "                    \"exons\": [{\"start\": location.start + 1, \"end\": location.end} for location in feature.location.parts],\n",
    "                    \"strand\": feature.location.strand\n",
    "                }\n",
    "            }\n",
    "            #if \"translation\" in feature.qualifiers:\n",
    "            #    extra_gene[\"translation\"] = feature.qualifiers[\"translation\"][0]\n",
    "            annots[\"extra_genes\"].append(extra_gene)\n",
    "            annot = {\n",
    "                \"id\": gene_id\n",
    "            }\n",
    "            if \"product\" in feature.qualifiers:\n",
    "                annot[\"product\"] = feature.qualifiers[\"product\"][0]\n",
    "            if \"note\" in feature.qualifiers:\n",
    "                annot[\"comments\"] = feature.qualifiers[\"note\"][0]\n",
    "            if len(annot.keys()) > 1:\n",
    "                annots[\"annotations\"].append(annot)\n",
    "\n",
    "    # check extra genes annotation\n",
    "    extra_genes = []\n",
    "    for i, extra_gene in enumerate(annots.get(\"extra_genes\", [])):\n",
    "        gene_id = extra_gene[\"id\"]\n",
    "        if gene_id.upper() not in [gid.upper() for gid in cluster_cds_ids] or gene_id.upper() in [gid.upper() for gid in \"/\".join(added_mibig_genes.get(bgc_id, \"\")).split(\"/\")]:\n",
    "            # check location, if outside the cluster, remove the locus information\n",
    "            if \"location\" in extra_gene:\n",
    "                exons = extra_gene[\"location\"].get(\"exons\", [])\n",
    "                if len(exons) > 0:\n",
    "                    cds_start = len(gbk_record.seq)\n",
    "                    cds_end = 0\n",
    "                    for exon in exons:\n",
    "                        if exon[\"start\"] < cds_start:\n",
    "                            cds_start = exon[\"start\"]\n",
    "                        if exon[\"end\"] > cds_end:\n",
    "                            cds_end = exon[\"end\"]\n",
    "                    if \"start_coord\" in loci and (cds_start < loci[\"start_coord\"] or cds_end > loci[\"end_coord\"]):\n",
    "                        extra_gene.pop(\"location\", None)\n",
    "                else:\n",
    "                    extra_gene.pop(\"location\", None)\n",
    "            extra_genes.append(extra_gene)\n",
    "            cluster_cds_ids.add(gene_id)\n",
    "        else:\n",
    "            if bgc_id not in removed_extra_genes:\n",
    "                removed_extra_genes[bgc_id] = []\n",
    "            removed_extra_genes[bgc_id].append(str(i))\n",
    "            \n",
    "    if len(extra_genes) > 0:\n",
    "        annots[\"extra_genes\"] = extra_genes\n",
    "    else:\n",
    "        annots.pop(\"extra_genes\", None)\n",
    "        \n",
    "    if (\"start_coord\" not in loci):\n",
    "        nt_length = len(gbk_record.seq)\n",
    "    else:\n",
    "        nt_length = loci[\"end_coord\"] - loci[\"start_coord\"]\n",
    "    gene_counts = len(cluster_cds.keys()) + len(extra_genes)\n",
    "    nt_length_and_cds_count[bgc_id] = (nt_length, gene_counts)\n",
    "            \n",
    "    # remove junk ids\n",
    "    cluster_cds_ids.discard(\"\")\n",
    "    cluster_cds_ids.discard(None)\n",
    "    \n",
    "    # check annotation, remove if no added info\n",
    "    gene_annots = []\n",
    "    for i, annot in enumerate(annots.get(\"annotations\", [])):\n",
    "        approve = False\n",
    "        found_id = None\n",
    "        for gid in cluster_cds_ids:\n",
    "            if annot.get(\"id\", \"\").upper() == gid.upper():\n",
    "                found_id = gid\n",
    "                annot[\"id\"] = gid\n",
    "                break\n",
    "            elif annot.get(\"name\", \"\").upper() == gid.upper():\n",
    "                found_id = gid\n",
    "                annot[\"name\"] = gid\n",
    "                break\n",
    "        ids = \"id='{}',name='{}'\".format(annot.get(\"id\", \"\"), annot.get(\"name\", \"\"))\n",
    "        if True:\n",
    "            if len(annot.get(\"name\", \"\")) < 1:\n",
    "                annot.pop(\"name\", None)\n",
    "            if len(annot.get(\"product\", \"\")) < 1:\n",
    "                annot.pop(\"product\", None)\n",
    "            if len(annot.get(\"mut_pheno\", \"\")) < 1:\n",
    "                annot.pop(\"mut_pheno\", None)\n",
    "            if len(annot.get(\"comments\", \"\")) < 1:\n",
    "                annot.pop(\"comments\", None)\n",
    "            g_functions = []\n",
    "            for g_function in annot.get(\"functions\", []):\n",
    "                if len(g_function.get(\"category\", \"\")) < 1:\n",
    "                    continue\n",
    "                elif len(g_function.get(\"evidence\", [])) < 1:\n",
    "                    continue\n",
    "                elif len(set(g_function[\"evidence\"]) - set([\"Sequence-based prediction\", \"Other in vivo study\", \"Heterologous expression\", \"Knock-out\", \"Activity assay\"])) > 0:\n",
    "                    continue\n",
    "                else:\n",
    "                    g_functions.append(g_function)\n",
    "            if len(g_functions) > 0:\n",
    "                annot[\"functions\"] = g_functions\n",
    "            else:\n",
    "                annot.pop(\"functions\", None)\n",
    "            if len(annot.get(\"tailoring\", [])) < 1:\n",
    "                annot.pop(\"tailoring\", None)\n",
    "            if len(annot.get(\"publications\", [])) < 1:\n",
    "                annot.pop(\"publications\", None)\n",
    "                \n",
    "            if found_id == None:\n",
    "                annot = {} # uncomment to add to extra genes when not found                \n",
    "                #if len(annot.get(\"id\", \"\")) < 1:\n",
    "                #    if len(annot.get(\"name\", \"\")) < 1:\n",
    "                #        annot = {}\n",
    "                #    else:\n",
    "                #        annot[\"id\"] = annot[\"name\"]\n",
    "                #if len(annot.get(\"id\", \"\")) > 0:\n",
    "                #    if \"extra_genes\" not in annots:\n",
    "                #        annots[\"extra_genes\"] = []\n",
    "                #    annots[\"extra_genes\"].append({\"id\": annot[\"id\"]})\n",
    "                #    cluster_cds_ids.add(annot[\"id\"])\n",
    "                    \n",
    "            if len(annot.keys()) > 1: # 1 is the id\n",
    "                approve = True\n",
    "        if approve:\n",
    "            gene_annots.append(annot)\n",
    "        else:\n",
    "            if bgc_id not in removed_annotations:\n",
    "                removed_annotations[bgc_id] = []\n",
    "            removed_annotations[bgc_id].append(ids)\n",
    "    if len(gene_annots) > 0:\n",
    "        annots[\"annotations\"] = gene_annots\n",
    "    else:\n",
    "        annots.pop(\"annotations\", None)\n",
    "                \n",
    "    # check operons\n",
    "    operons = []\n",
    "    for i, operon in enumerate(annots.get(\"operons\", [])):\n",
    "        o_genes = set([gid.upper() for gid in operon.get(\"genes\", [])])\n",
    "        o_evidence = set(operon.get(\"evidence\", [])).intersection(set([\"Sequence-based prediction\", \"RACE\", \"ChIPseq\", \"RNAseq\"]))\n",
    "        if len(o_genes) > 0 and len(o_genes - cluster_cds_ids) < 1:\n",
    "            if len(o_evidence) > 0:\n",
    "                operons.append({\n",
    "                    \"genes\": list(o_genes),\n",
    "                    \"evidence\": list(o_evidence)\n",
    "                })\n",
    "                continue\n",
    "        if bgc_id not in removed_operons:\n",
    "            removed_operons[bgc_id] = []\n",
    "        removed_operons[bgc_id].append(str(i))\n",
    "        \n",
    "    if len(operons) > 0:\n",
    "        annots[\"operons\"] = operons\n",
    "    else:\n",
    "        annots.pop(\"operons\", None)\n",
    "\n",
    "    # try to fix ripp leader/follower sequences\n",
    "    if \"ripp\" in data[\"cluster\"]:\n",
    "        fixed_precursors = []\n",
    "        unfixable_precursors = []\n",
    "        for i, precursor in enumerate(data[\"cluster\"][\"ripp\"].get(\"precursor_genes\", [])):\n",
    "            if \"gene_id\" in precursor:\n",
    "                cds_feature = None\n",
    "                for gid in cluster_cds_ids:\n",
    "                    if precursor[\"gene_id\"].upper() == gid.upper():\n",
    "                        cds_feature = cluster_cds[gid]\n",
    "                        break\n",
    "                if cds_feature != None:\n",
    "                    if \"translation\" in cds_feature.qualifiers:\n",
    "                        aa_sequence = str(cds_feature.qualifiers[\"translation\"][0])\n",
    "                        leader_length = precursor.get(\"leader_sequence\", 0)\n",
    "                        follower_length = precursor.get(\"follower_sequence\", 0)\n",
    "                        if leader_length < 1 or leader_length > len(aa_sequence) or follower_length > len(aa_sequence):\n",
    "                            print(\"[ripp] Wrong leader/follower length information {}\".format(precursor[\"gene_id\"]))\n",
    "                            unfixable_precursors.append(i)\n",
    "                        elif len(precursor.get(\"core_sequence\", [])) < 1:\n",
    "                            print(\"[ripp] No core sequence {}\".format(precursor[\"gene_id\"]))\n",
    "                            unfixable_precursors.append(i)\n",
    "                        else:\n",
    "                            cores_checked = True\n",
    "                            for core in precursor[\"core_sequence\"]:\n",
    "                                if len(core) < 1:\n",
    "                                    cores_checked = False\n",
    "                                    break\n",
    "                                core_start = aa_sequence.find(core)\n",
    "                                if core_start < 0:\n",
    "                                    # can't find core\n",
    "                                    cores_checked = False\n",
    "                                    break\n",
    "                                elif core_start < leader_length or core_start + len(core) >= len(aa_sequence) - follower_length:\n",
    "                                    # core overlaps with leader/follower\n",
    "                                    cores_checked = False\n",
    "                                    break\n",
    "                            if cores_checked:\n",
    "                                precursor[\"leader_sequence\"] = aa_sequence[0:leader_length]\n",
    "                                if follower_length > 0:\n",
    "                                    precursor[\"follower_sequence\"] = aa_sequence[follower_length - 1:]\n",
    "                                else:\n",
    "                                    precursor.pop(\"follower_sequence\", None)\n",
    "                                fixed_precursors.append(i)\n",
    "                            else:\n",
    "                                print(\"[ripp] Wrong sequence information for {}\".format(precursor[\"gene_id\"]))\n",
    "                                unfixable_precursors.append(i)\n",
    "                    else:\n",
    "                        print(\"[ripp] Can't find AA sequence for {}\".format(precursor[\"gene_id\"]))\n",
    "                        unfixable_precursors.append(i)\n",
    "                else:\n",
    "                    print(\"[ripp] Can't find CDS feature for {}\".format(precursor[\"gene_id\"]))\n",
    "                    unfixable_precursors.append(i)\n",
    "                    \n",
    "        if len(fixed_precursors) > 0 and len(fixed_precursors) == len(data[\"cluster\"][\"ripp\"][\"precursor_genes\"]):\n",
    "            # remove the retired flag from ripp.leader/follower_sequence\n",
    "            print(\"[ripp] solved precursors {}\".format(bgc_id))\n",
    "            if bgc_id not in fixed:\n",
    "                fixed[bgc_id] = []\n",
    "            fixed[bgc_id].append(\"ripp.leader/follower_sequence\")\n",
    "        elif len(unfixable_precursors) > 0:\n",
    "            ripps_to_fix[bgc_id] = unfixable_precursors\n",
    "            # temporarily remove the precursors, and remove the retired flag from ripp.leader/follower_sequence\n",
    "            precursors = []\n",
    "            for i, precursor in enumerate(data[\"cluster\"][\"ripp\"][\"precursor_genes\"]):\n",
    "                if i not in unfixable_precursors:\n",
    "                    precursors.append(precursor)\n",
    "            if len(precursors) > 0:\n",
    "                data[\"cluster\"][\"ripp\"][\"precursor_genes\"] = precursors\n",
    "            else:\n",
    "                data[\"cluster\"][\"ripp\"].pop(\"precursor_genes\", None)\n",
    "            if bgc_id not in fixed:\n",
    "                fixed[bgc_id] = []\n",
    "            fixed[bgc_id].append(\"ripp.leader/follower_sequence\")            \n",
    "            \n",
    "        \n",
    "    # update gene ids in specific_infos\n",
    "    data = make_gene_id_match_reference(data, cluster_cds_ids)\n",
    "        \n",
    "    if len(annots.keys()) > 0:\n",
    "        data[\"cluster\"][\"genes\"] = annots\n",
    "    else:\n",
    "        data[\"cluster\"].pop(\"genes\", None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_data(data, validator, schema_props):\n",
    "    for error in sorted(validator.iter_errors(data), key=str):\n",
    "        try:\n",
    "            if error.path[-2] == \"evidence\" or error.path[-1] == \"evidence\":\n",
    "                if error.path[-3] == \"loci\" or error.path[-2] == \"loci\":\n",
    "                    continue\n",
    "            elif error.path[-1] == \"nr_iterations\":\n",
    "                continue\n",
    "            elif error.path[-1] == \"module_number\":\n",
    "                continue\n",
    "            elif error.path[-2] == \"proteinogenic\":\n",
    "                continue\n",
    "            elif error.path[-1] in [\"leader_sequence\", \"follower_sequence\"]:\n",
    "                continue\n",
    "            elif error.path[-1] == \"organism\":\n",
    "                if data[\"cluster\"][\"mibig_accession\"] in no_gbk:\n",
    "                    continue\n",
    "            elif error.path[-1] == \"subcluster\" and error.path[-3] == \"chem_moieties\":\n",
    "                continue\n",
    "        except:\n",
    "            pass\n",
    "        print(error.message)\n",
    "        #sys.exit(0)\n",
    "    this_file_props = count_props(data, \"\", {})\n",
    "    for prop in this_file_props:\n",
    "        if prop not in schema_props.keys():\n",
    "            print(prop)\n",
    "            sys.exit(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0000001\n",
      "Scanning BGC0000002\n",
      "Scanning BGC0000003\n",
      "Scanning BGC0000004\n",
      "Scanning BGC0000005\n",
      "Scanning BGC0000006\n",
      "Scanning BGC0000007\n",
      "Scanning BGC0000008\n",
      "Scanning BGC0000009\n",
      "Scanning BGC0000010\n",
      "Scanning BGC0000011\n",
      "Scanning BGC0000012\n",
      "Scanning BGC0000013\n",
      "Scanning BGC0000014\n",
      "Scanning BGC0000015\n",
      "Scanning BGC0000016\n",
      "Scanning BGC0000017\n",
      "Scanning BGC0000018\n",
      "Scanning BGC0000019\n",
      "Scanning BGC0000020\n",
      "Scanning BGC0000021\n",
      "Scanning BGC0000022\n",
      "Scanning BGC0000023\n",
      "Scanning BGC0000024\n",
      "Scanning BGC0000025\n",
      "Scanning BGC0000026\n",
      "Scanning BGC0000027\n",
      "Scanning BGC0000028\n",
      "Scanning BGC0000029\n",
      "Scanning BGC0000030\n",
      "Scanning BGC0000031\n",
      "Scanning BGC0000032\n",
      "Scanning BGC0000033\n",
      "Scanning BGC0000034\n",
      "Scanning BGC0000035\n",
      "Scanning BGC0000036\n",
      "Scanning BGC0000037\n",
      "Scanning BGC0000038\n",
      "Scanning BGC0000039\n",
      "Scanning BGC0000040\n",
      "Scanning BGC0000041\n",
      "Scanning BGC0000042\n",
      "Scanning BGC0000043\n",
      "Scanning BGC0000044\n",
      "Scanning BGC0000045\n",
      "Scanning BGC0000046\n",
      "Scanning BGC0000047\n",
      "Scanning BGC0000048\n",
      "Scanning BGC0000049\n",
      "Scanning BGC0000050\n",
      "Scanning BGC0000051\n",
      "Scanning BGC0000052\n",
      "Scanning BGC0000053\n",
      "Scanning BGC0000054\n",
      "Scanning BGC0000055\n",
      "Scanning BGC0000056\n",
      "Scanning BGC0000057\n",
      "Scanning BGC0000058\n",
      "Scanning BGC0000059\n",
      "Scanning BGC0000060\n",
      "Scanning BGC0000061\n",
      "Scanning BGC0000062\n",
      "Scanning BGC0000063\n",
      "Scanning BGC0000064\n",
      "Scanning BGC0000065\n",
      "Scanning BGC0000066\n",
      "Scanning BGC0000067\n",
      "Scanning BGC0000068\n",
      "Scanning BGC0000069\n",
      "Scanning BGC0000070\n",
      "Scanning BGC0000072\n",
      "Scanning BGC0000073\n",
      "Scanning BGC0000074\n",
      "Scanning BGC0000075\n",
      "Scanning BGC0000076\n",
      "Scanning BGC0000077\n",
      "Scanning BGC0000078\n",
      "Scanning BGC0000079\n",
      "Scanning BGC0000080\n",
      "Scanning BGC0000081\n",
      "Scanning BGC0000082\n",
      "Scanning BGC0000083\n",
      "Scanning BGC0000084\n",
      "Scanning BGC0000085\n",
      "Scanning BGC0000086\n",
      "Scanning BGC0000087\n",
      "Scanning BGC0000088\n",
      "Scanning BGC0000089\n",
      "Scanning BGC0000090\n",
      "Scanning BGC0000091\n",
      "Scanning BGC0000092\n",
      "Scanning BGC0000093\n",
      "Scanning BGC0000094\n",
      "Scanning BGC0000095\n",
      "Scanning BGC0000096\n",
      "Scanning BGC0000097\n",
      "Scanning BGC0000098\n",
      "Scanning BGC0000099\n",
      "Scanning BGC0000100\n",
      "Scanning BGC0000101\n",
      "Scanning BGC0000102\n",
      "Scanning BGC0000103\n",
      "Scanning BGC0000104\n",
      "Scanning BGC0000105\n",
      "Scanning BGC0000106\n",
      "Scanning BGC0000107\n",
      "Scanning BGC0000108\n",
      "Scanning BGC0000109\n",
      "Scanning BGC0000110\n",
      "Scanning BGC0000111\n",
      "Scanning BGC0000112\n",
      "Scanning BGC0000113\n",
      "Scanning BGC0000114\n",
      "Scanning BGC0000115\n",
      "Scanning BGC0000116\n",
      "Scanning BGC0000117\n",
      "Scanning BGC0000118\n",
      "Scanning BGC0000119\n",
      "Scanning BGC0000120\n",
      "Scanning BGC0000121\n",
      "Scanning BGC0000122\n",
      "Scanning BGC0000123\n",
      "Scanning BGC0000124\n",
      "Scanning BGC0000125\n",
      "Scanning BGC0000126\n",
      "Scanning BGC0000127\n",
      "Scanning BGC0000128\n",
      "Scanning BGC0000129\n",
      "Scanning BGC0000130\n",
      "Scanning BGC0000131\n",
      "Scanning BGC0000132\n",
      "Scanning BGC0000133\n",
      "Scanning BGC0000134\n",
      "Scanning BGC0000135\n",
      "Scanning BGC0000136\n",
      "Scanning BGC0000137\n",
      "Scanning BGC0000138\n",
      "Scanning BGC0000140\n",
      "Scanning BGC0000141\n",
      "Scanning BGC0000142\n",
      "Scanning BGC0000143\n",
      "Scanning BGC0000144\n",
      "Scanning BGC0000145\n",
      "Scanning BGC0000146\n",
      "Scanning BGC0000147\n",
      "Scanning BGC0000148\n",
      "Scanning BGC0000149\n",
      "Scanning BGC0000150\n",
      "Scanning BGC0000151\n",
      "Scanning BGC0000152\n",
      "Scanning BGC0000153\n",
      "Scanning BGC0000154\n",
      "Scanning BGC0000155\n",
      "Scanning BGC0000156\n",
      "Scanning BGC0000157\n",
      "Scanning BGC0000158\n",
      "Scanning BGC0000159\n",
      "Scanning BGC0000160\n",
      "Scanning BGC0000161\n",
      "Scanning BGC0000162\n",
      "Scanning BGC0000163\n",
      "Scanning BGC0000164\n",
      "Scanning BGC0000165\n",
      "Scanning BGC0000166\n",
      "Scanning BGC0000167\n",
      "Scanning BGC0000168\n",
      "Scanning BGC0000170\n",
      "Scanning BGC0000171\n",
      "Scanning BGC0000172\n",
      "Scanning BGC0000173\n",
      "Scanning BGC0000174\n",
      "Scanning BGC0000175\n",
      "Scanning BGC0000176\n",
      "Scanning BGC0000177\n",
      "Scanning BGC0000178\n",
      "Scanning BGC0000179\n",
      "Scanning BGC0000180\n",
      "Scanning BGC0000181\n",
      "Scanning BGC0000182\n",
      "Scanning BGC0000183\n",
      "Scanning BGC0000184\n",
      "Scanning BGC0000185\n",
      "Scanning BGC0000186\n",
      "Scanning BGC0000187\n",
      "Scanning BGC0000188\n",
      "Scanning BGC0000189\n",
      "Scanning BGC0000190\n",
      "Scanning BGC0000191\n",
      "Scanning BGC0000192\n",
      "Scanning BGC0000193\n",
      "Scanning BGC0000194\n",
      "Scanning BGC0000195\n",
      "Scanning BGC0000196\n",
      "Scanning BGC0000197\n",
      "Scanning BGC0000198\n",
      "Scanning BGC0000199\n",
      "Scanning BGC0000200\n",
      "Scanning BGC0000201\n",
      "Scanning BGC0000202\n",
      "Scanning BGC0000203\n",
      "Scanning BGC0000204\n",
      "Scanning BGC0000205\n",
      "Scanning BGC0000206\n",
      "Scanning BGC0000207\n",
      "Scanning BGC0000208\n",
      "Scanning BGC0000209\n",
      "Scanning BGC0000210\n",
      "Scanning BGC0000211\n",
      "Scanning BGC0000212\n",
      "Scanning BGC0000213\n",
      "Scanning BGC0000214\n",
      "Scanning BGC0000215\n",
      "Scanning BGC0000216\n",
      "Scanning BGC0000217\n",
      "Scanning BGC0000218\n",
      "Scanning BGC0000219\n",
      "Scanning BGC0000220\n",
      "Scanning BGC0000221\n",
      "Scanning BGC0000222\n",
      "Scanning BGC0000223\n",
      "Scanning BGC0000224\n",
      "Scanning BGC0000225\n",
      "Scanning BGC0000226\n",
      "Scanning BGC0000227\n",
      "Scanning BGC0000228\n",
      "Scanning BGC0000229\n",
      "Scanning BGC0000230\n",
      "Scanning BGC0000231\n",
      "Scanning BGC0000232\n",
      "Scanning BGC0000233\n",
      "Scanning BGC0000234\n",
      "Scanning BGC0000235\n",
      "Scanning BGC0000236\n",
      "Scanning BGC0000237\n",
      "Scanning BGC0000238\n",
      "Scanning BGC0000239\n",
      "Scanning BGC0000240\n",
      "Scanning BGC0000241\n",
      "Scanning BGC0000242\n",
      "Scanning BGC0000243\n",
      "Scanning BGC0000244\n",
      "Scanning BGC0000245\n",
      "Scanning BGC0000246\n",
      "Scanning BGC0000247\n",
      "Scanning BGC0000248\n",
      "Scanning BGC0000249\n",
      "Scanning BGC0000250\n",
      "Scanning BGC0000251\n",
      "Scanning BGC0000252\n",
      "Scanning BGC0000253\n",
      "Scanning BGC0000254\n",
      "Scanning BGC0000255\n",
      "['629395', '2']\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0000256\n",
      "Scanning BGC0000257\n",
      "Scanning BGC0000258\n",
      "Scanning BGC0000259\n",
      "Scanning BGC0000260\n",
      "Scanning BGC0000261\n",
      "Scanning BGC0000262\n",
      "Scanning BGC0000263\n",
      "Scanning BGC0000264\n",
      "Scanning BGC0000265\n",
      "Scanning BGC0000266\n",
      "Scanning BGC0000267\n",
      "Scanning BGC0000268\n",
      "Scanning BGC0000269\n",
      "Scanning BGC0000270\n",
      "Scanning BGC0000271\n",
      "Scanning BGC0000272\n",
      "Scanning BGC0000273\n",
      "Scanning BGC0000274\n",
      "Scanning BGC0000275\n",
      "Scanning BGC0000276\n",
      "Scanning BGC0000277\n",
      "Scanning BGC0000278\n",
      "Scanning BGC0000279\n",
      "Scanning BGC0000280\n",
      "Scanning BGC0000281\n",
      "Scanning BGC0000282\n",
      "Scanning BGC0000283\n",
      "Scanning BGC0000284\n",
      "Scanning BGC0000285\n",
      "Scanning BGC0000286\n",
      "Scanning BGC0000287\n",
      "Scanning BGC0000288\n",
      "Scanning BGC0000289\n",
      "Scanning BGC0000290\n",
      "Scanning BGC0000291\n",
      "Scanning BGC0000292\n",
      "Scanning BGC0000293\n",
      "Scanning BGC0000294\n",
      "Scanning BGC0000295\n",
      "Scanning BGC0000296\n",
      "Scanning BGC0000297\n",
      "Scanning BGC0000298\n",
      "Scanning BGC0000299\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0000300\n",
      "Scanning BGC0000301\n",
      "Scanning BGC0000302\n",
      "Scanning BGC0000303\n",
      "Scanning BGC0000304\n",
      "Scanning BGC0000305\n",
      "Scanning BGC0000306\n",
      "Scanning BGC0000307\n",
      "Scanning BGC0000308\n",
      "Scanning BGC0000309\n",
      "Scanning BGC0000310\n",
      "Scanning BGC0000311\n",
      "Scanning BGC0000312\n",
      "Scanning BGC0000313\n",
      "Scanning BGC0000314\n",
      "Scanning BGC0000315\n",
      "Scanning BGC0000316\n",
      "Scanning BGC0000317\n",
      "Scanning BGC0000318\n",
      "Scanning BGC0000319\n",
      "Scanning BGC0000320\n",
      "Scanning BGC0000321\n",
      "Scanning BGC0000322\n",
      "Scanning BGC0000323\n",
      "Scanning BGC0000324\n",
      "Scanning BGC0000325\n",
      "Scanning BGC0000326\n",
      "Scanning BGC0000327\n",
      "Scanning BGC0000328\n",
      "Scanning BGC0000329\n",
      "Scanning BGC0000330\n",
      "Scanning BGC0000331\n",
      "Scanning BGC0000332\n",
      "Scanning BGC0000333\n",
      "Scanning BGC0000334\n",
      "Scanning BGC0000335\n",
      "Scanning BGC0000336\n",
      "Scanning BGC0000337\n",
      "Scanning BGC0000338\n",
      "Scanning BGC0000339\n",
      "Scanning BGC0000340\n",
      "Scanning BGC0000341\n",
      "Scanning BGC0000342\n",
      "Scanning BGC0000343\n",
      "Scanning BGC0000344\n",
      "Scanning BGC0000345\n",
      "Scanning BGC0000346\n",
      "Scanning BGC0000347\n",
      "Scanning BGC0000348\n",
      "Scanning BGC0000349\n",
      "Scanning BGC0000350\n",
      "Scanning BGC0000351\n",
      "Scanning BGC0000352\n",
      "Scanning BGC0000353\n",
      "Scanning BGC0000354\n",
      "Scanning BGC0000355\n",
      "Scanning BGC0000356\n",
      "Scanning BGC0000357\n",
      "Scanning BGC0000358\n",
      "Scanning BGC0000359\n",
      "Scanning BGC0000360\n",
      "Scanning BGC0000361\n",
      "Scanning BGC0000362\n",
      "Scanning BGC0000363\n",
      "Scanning BGC0000364\n",
      "Scanning BGC0000365\n",
      "Scanning BGC0000366\n",
      "Scanning BGC0000367\n",
      "Scanning BGC0000368\n",
      "Scanning BGC0000369\n",
      "Scanning BGC0000370\n",
      "Scanning BGC0000371\n",
      "Scanning BGC0000372\n",
      "Scanning BGC0000373\n",
      "Scanning BGC0000374\n",
      "Scanning BGC0000375\n",
      "Scanning BGC0000376\n",
      "Scanning BGC0000377\n",
      "Scanning BGC0000378\n",
      "Scanning BGC0000379\n",
      "Scanning BGC0000380\n",
      "Scanning BGC0000381\n",
      "Scanning BGC0000382\n",
      "Scanning BGC0000383\n",
      "Scanning BGC0000384\n",
      "Scanning BGC0000385\n",
      "Scanning BGC0000386\n",
      "Scanning BGC0000387\n",
      "Scanning BGC0000388\n",
      "Scanning BGC0000389\n",
      "Scanning BGC0000391\n",
      "Scanning BGC0000392\n",
      "Scanning BGC0000393\n",
      "Scanning BGC0000394\n",
      "Scanning BGC0000395\n",
      "Scanning BGC0000396\n",
      "Scanning BGC0000397\n",
      "Scanning BGC0000398\n",
      "Scanning BGC0000399\n",
      "Scanning BGC0000400\n",
      "Scanning BGC0000401\n",
      "Scanning BGC0000402\n",
      "Scanning BGC0000403\n",
      "Scanning BGC0000404\n",
      "Scanning BGC0000405\n",
      "Scanning BGC0000406\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0000407\n",
      "Scanning BGC0000408\n",
      "Scanning BGC0000409\n",
      "Scanning BGC0000410\n",
      "Scanning BGC0000411\n",
      "Scanning BGC0000412\n",
      "Scanning BGC0000413\n",
      "Scanning BGC0000414\n",
      "Scanning BGC0000415\n",
      "Scanning BGC0000416\n",
      "Scanning BGC0000417\n",
      "Scanning BGC0000418\n",
      "Scanning BGC0000419\n",
      "Scanning BGC0000420\n",
      "Scanning BGC0000421\n",
      "Scanning BGC0000422\n",
      "Scanning BGC0000423\n",
      "Scanning BGC0000424\n",
      "Scanning BGC0000425\n",
      "Scanning BGC0000426\n",
      "Scanning BGC0000427\n",
      "Scanning BGC0000428\n",
      "Scanning BGC0000429\n",
      "Scanning BGC0000430\n",
      "Scanning BGC0000431\n",
      "Scanning BGC0000432\n",
      "Scanning BGC0000433\n",
      "Scanning BGC0000434\n",
      "Scanning BGC0000435\n",
      "Scanning BGC0000436\n",
      "Scanning BGC0000437\n",
      "Scanning BGC0000438\n",
      "Scanning BGC0000439\n",
      "Scanning BGC0000440\n",
      "Scanning BGC0000441\n",
      "Scanning BGC0000442\n",
      "Scanning BGC0000443\n",
      "Scanning BGC0000444\n",
      "Scanning BGC0000445\n",
      "Scanning BGC0000446\n",
      "Scanning BGC0000447\n",
      "Scanning BGC0000448\n",
      "Scanning BGC0000449\n",
      "Scanning BGC0000450\n",
      "Scanning BGC0000451\n",
      "Scanning BGC0000452\n",
      "Scanning BGC0000453\n",
      "Scanning BGC0000454\n",
      "Scanning BGC0000455\n",
      "Scanning BGC0000456\n",
      "Scanning BGC0000457\n",
      "Scanning BGC0000458\n",
      "Scanning BGC0000459\n",
      "Scanning BGC0000460\n",
      "Scanning BGC0000461\n",
      "Scanning BGC0000462\n",
      "Scanning BGC0000463\n",
      "Scanning BGC0000464\n",
      "Scanning BGC0000465\n",
      "Scanning BGC0000466\n",
      "Scanning BGC0000467\n",
      "Scanning BGC0000468\n",
      "[ripp] Can't find CDS feature for bmbc,ccm09443.1\n",
      "Scanning BGC0000469\n",
      "Scanning BGC0000470\n",
      "Scanning BGC0000471\n",
      "Scanning BGC0000472\n",
      "[ripp] Can't find CDS feature for ada00388\n",
      "Scanning BGC0000473\n",
      "[ripp] Can't find CDS feature for cao82085\n",
      "Scanning BGC0000474\n",
      "[ripp] Can't find CDS feature for cap64339\n",
      "Scanning BGC0000475\n",
      "Scanning BGC0000476\n",
      "[ripp] Wrong sequence information for abi99461.1\n",
      "Scanning BGC0000477\n",
      "[ripp] solved precursors BGC0000477\n",
      "Scanning BGC0000478\n",
      "[ripp] Can't find CDS feature for eu290743\n",
      "Scanning BGC0000479\n",
      "[ripp] solved precursors BGC0000479\n",
      "Scanning BGC0000480\n",
      "[ripp] solved precursors BGC0000480\n",
      "Scanning BGC0000481\n",
      "Scanning BGC0000482\n",
      "[ripp] Can't find CDS feature for eu290744\n",
      "Scanning BGC0000483\n",
      "Scanning BGC0000484\n",
      "Scanning BGC0000485\n",
      "Scanning BGC0000486\n",
      "Scanning BGC0000487\n",
      "Scanning BGC0000488\n",
      "[ripp] Wrong sequence information for CAD97580.1\n",
      "Scanning BGC0000489\n",
      "Scanning BGC0000490\n",
      "Scanning BGC0000491\n",
      "Scanning BGC0000492\n",
      "Scanning BGC0000493\n",
      "Scanning BGC0000494\n",
      "Scanning BGC0000495\n",
      "Scanning BGC0000496\n",
      "Scanning BGC0000497\n",
      "Scanning BGC0000498\n",
      "Scanning BGC0000499\n",
      "Scanning BGC0000500\n",
      "Scanning BGC0000501\n",
      "[ripp] Can't find CDS feature for yp_003114945.1\n",
      "Scanning BGC0000502\n",
      "Scanning BGC0000503\n",
      "[ripp] solved precursors BGC0000503\n",
      "Scanning BGC0000504\n",
      "Scanning BGC0000505\n",
      "Scanning BGC0000506\n",
      "[ripp] Wrong sequence information for etns\n",
      "Scanning BGC0000507\n",
      "Scanning BGC0000508\n",
      "Scanning BGC0000509\n",
      "[ripp] Wrong sequence information for afn69432.1\n",
      "Scanning BGC0000510\n",
      "Scanning BGC0000511\n",
      "[ripp] Can't find CDS feature for eris\n",
      "Scanning BGC0000513\n",
      "Scanning BGC0000514\n",
      "Scanning BGC0000515\n",
      "[ripp] Can't find CDS feature for \n",
      "Scanning BGC0000516\n",
      "Scanning BGC0000517\n",
      "[ripp] Wrong sequence information for BAB04172.1\n",
      "Scanning BGC0000518\n",
      "Scanning BGC0000519\n",
      "[ripp] Can't find CDS feature for c0mp60\n",
      "[ripp] Can't find CDS feature for c0mp59\n",
      "Scanning BGC0000520\n",
      "Scanning BGC0000521\n",
      "Scanning BGC0000522\n",
      "Scanning BGC0000523\n",
      "Scanning BGC0000525\n",
      "Scanning BGC0000526\n",
      "Scanning BGC0000527\n",
      "Scanning BGC0000528\n",
      "Scanning BGC0000529\n",
      "[ripp] Can't find CDS feature for adk32554\n",
      "Scanning BGC0000530\n",
      "Scanning BGC0000531\n",
      "Scanning BGC0000532\n",
      "Scanning BGC0000533\n",
      "Scanning BGC0000534\n",
      "Scanning BGC0000535\n",
      "[ripp] solved precursors BGC0000535\n",
      "Scanning BGC0000536\n",
      "Scanning BGC0000537\n",
      "Scanning BGC0000538\n",
      "Scanning BGC0000539\n",
      "Scanning BGC0000540\n",
      "Scanning BGC0000541\n",
      "Scanning BGC0000542\n",
      "Scanning BGC0000543\n",
      "Scanning BGC0000544\n",
      "[ripp] Can't find CDS feature for ccq18695\n",
      "Scanning BGC0000545\n",
      "Scanning BGC0000546\n",
      "[ripp] Can't find CDS feature for sshg_03589,wp_003950016.1\n",
      "Scanning BGC0000547\n",
      "Scanning BGC0000548\n",
      "Scanning BGC0000549\n",
      "Scanning BGC0000550\n",
      "Scanning BGC0000551\n",
      "[ripp] Can't find CDS feature for rams\n",
      "Scanning BGC0000552\n",
      "Scanning BGC0000553\n",
      "[ripp] solved precursors BGC0000553\n",
      "Scanning BGC0000554\n",
      "[ripp] Can't find CDS feature for ssgg_rs13165,wp_006126649.1\n",
      "Scanning BGC0000555\n",
      "Scanning BGC0000556\n",
      "Scanning BGC0000557\n",
      "Scanning BGC0000558\n",
      "Scanning BGC0000559\n",
      "[ripp] Wrong sequence information for spas\n",
      "Scanning BGC0000560\n",
      "Scanning BGC0000561\n",
      "Scanning BGC0000562\n",
      "Scanning BGC0000563\n",
      "[ripp] solved precursors BGC0000563\n",
      "Scanning BGC0000564\n",
      "[ripp] Wrong leader/follower length information saga\n",
      "Scanning BGC0000565\n",
      "Scanning BGC0000566\n",
      "[ripp] solved precursors BGC0000566\n",
      "Scanning BGC0000567\n",
      "Scanning BGC0000568\n",
      "Scanning BGC0000569\n",
      "[ripp] Can't find CDS feature for cbj61635\n",
      "Scanning BGC0000570\n",
      "Scanning BGC0000571\n",
      "Scanning BGC0000572\n",
      "Scanning BGC0000573\n",
      "Scanning BGC0000574\n",
      "Scanning BGC0000575\n",
      "Scanning BGC0000576\n",
      "Scanning BGC0000577\n",
      "Scanning BGC0000578\n",
      "[ripp] Can't find CDS feature for ssgg_03858\n",
      "Scanning BGC0000579\n",
      "[ripp] Wrong sequence information for sseg_05172\n",
      "Scanning BGC0000580\n",
      "Scanning BGC0000581\n",
      "Scanning BGC0000582\n",
      "[ripp] solved precursors BGC0000582\n",
      "Scanning BGC0000583\n",
      "[ripp] Can't find CDS feature for grma\n",
      "Scanning BGC0000584\n",
      "Scanning BGC0000585\n",
      "Scanning BGC0000586\n",
      "Scanning BGC0000587\n",
      "Scanning BGC0000588\n",
      "Scanning BGC0000589\n",
      "Scanning BGC0000590\n",
      "Scanning BGC0000591\n",
      "Scanning BGC0000592\n",
      "[ripp] Can't find CDS feature for caq16116,mdna\n",
      "Scanning BGC0000593\n",
      "[ripp] Can't find CDS feature for caq16121\n",
      "Scanning BGC0000594\n",
      "Scanning BGC0000595\n",
      "Scanning BGC0000596\n",
      "Scanning BGC0000597\n",
      "[ripp] Can't find CDS feature for \n",
      "Scanning BGC0000598\n",
      "['629395', '2']\n",
      "[ripp] Can't find CDS feature for afs60639\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0000599\n",
      "Scanning BGC0000600\n",
      "Scanning BGC0000601\n",
      "[ripp] Wrong sequence information for cab11984.1\n",
      "Scanning BGC0000602\n",
      "Scanning BGC0000603\n",
      "[ripp] Wrong sequence information for acs50131.1\n",
      "Scanning BGC0000604\n",
      "[ripp] Wrong sequence information for acs83766.1\n",
      "Scanning BGC0000605\n",
      "Scanning BGC0000606\n",
      "Scanning BGC0000607\n",
      "Scanning BGC0000608\n",
      "Scanning BGC0000609\n",
      "Scanning BGC0000610\n",
      "Scanning BGC0000611\n",
      "Scanning BGC0000612\n",
      "[ripp] Can't find CDS feature for AAP11957\n",
      "[ripp] Can't find CDS feature for AAP11958\n",
      "[ripp] Can't find CDS feature for AAP11959\n",
      "[ripp] Can't find CDS feature for AAP11960\n",
      "Scanning BGC0000613\n",
      "[ripp] Wrong sequence information for acs83781.1\n",
      "Scanning BGC0000614\n",
      "Scanning BGC0000615\n",
      "[ripp] Wrong sequence information for ado67770.1\n",
      "Scanning BGC0000616\n",
      "[ripp] Wrong sequence information for rbam_029230\n",
      "Scanning BGC0000617\n",
      "Scanning BGC0000618\n",
      "Scanning BGC0000619\n",
      "Scanning BGC0000620\n",
      "Scanning BGC0000621\n",
      "Scanning BGC0000622\n",
      "Scanning BGC0000623\n",
      "Scanning BGC0000624\n",
      "Scanning BGC0000625\n",
      "Scanning BGC0000626\n",
      "Scanning BGC0000627\n",
      "Scanning BGC0000628\n",
      "[ripp] Wrong sequence information for hmpref0514_11728\n",
      "Scanning BGC0000629\n",
      "Scanning BGC0000630\n",
      "Scanning BGC0000631\n",
      "Scanning BGC0000632\n",
      "Scanning BGC0000633\n",
      "Scanning BGC0000634\n",
      "Scanning BGC0000635\n",
      "Scanning BGC0000636\n",
      "Scanning BGC0000637\n",
      "Scanning BGC0000638\n",
      "Scanning BGC0000639\n",
      "Scanning BGC0000640\n",
      "Scanning BGC0000641\n",
      "Scanning BGC0000642\n",
      "Scanning BGC0000643\n",
      "Scanning BGC0000644\n",
      "Scanning BGC0000645\n",
      "Scanning BGC0000646\n",
      "Scanning BGC0000647\n",
      "Scanning BGC0000648\n",
      "Scanning BGC0000649\n",
      "Scanning BGC0000650\n",
      "Scanning BGC0000651\n",
      "Scanning BGC0000652\n",
      "Scanning BGC0000653\n",
      "Scanning BGC0000654\n",
      "Scanning BGC0000655\n",
      "Scanning BGC0000656\n",
      "Scanning BGC0000657\n",
      "Scanning BGC0000658\n",
      "Scanning BGC0000659\n",
      "Scanning BGC0000660\n",
      "Scanning BGC0000661\n",
      "Scanning BGC0000662\n",
      "Scanning BGC0000663\n",
      "Scanning BGC0000664\n",
      "Scanning BGC0000665\n",
      "Scanning BGC0000666\n",
      "Scanning BGC0000667\n",
      "Scanning BGC0000668\n",
      "Scanning BGC0000669\n",
      "Scanning BGC0000670\n",
      "Scanning BGC0000671\n",
      "Scanning BGC0000672\n",
      "Scanning BGC0000673\n",
      "Scanning BGC0000674\n",
      "Scanning BGC0000675\n",
      "Scanning BGC0000676\n",
      "Scanning BGC0000677\n",
      "Scanning BGC0000678\n",
      "Scanning BGC0000679\n",
      "Scanning BGC0000680\n",
      "Scanning BGC0000682\n",
      "Scanning BGC0000683\n",
      "Scanning BGC0000684\n",
      "Scanning BGC0000685\n",
      "Scanning BGC0000686\n",
      "Scanning BGC0000687\n",
      "Scanning BGC0000688\n",
      "Scanning BGC0000689\n",
      "Scanning BGC0000690\n",
      "Scanning BGC0000691\n",
      "Scanning BGC0000692\n",
      "Scanning BGC0000693\n",
      "Scanning BGC0000694\n",
      "Scanning BGC0000695\n",
      "Scanning BGC0000696\n",
      "Scanning BGC0000697\n",
      "Scanning BGC0000698\n",
      "Scanning BGC0000699\n",
      "Scanning BGC0000700\n",
      "Scanning BGC0000701\n",
      "Scanning BGC0000702\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0000703\n",
      "Scanning BGC0000704\n",
      "Scanning BGC0000705\n",
      "Scanning BGC0000706\n",
      "Scanning BGC0000707\n",
      "Scanning BGC0000708\n",
      "Scanning BGC0000709\n",
      "Scanning BGC0000710\n",
      "Scanning BGC0000711\n",
      "Scanning BGC0000712\n",
      "Scanning BGC0000713\n",
      "Scanning BGC0000714\n",
      "Scanning BGC0000715\n",
      "Scanning BGC0000716\n",
      "Scanning BGC0000717\n",
      "Scanning BGC0000718\n",
      "Scanning BGC0000719\n",
      "Scanning BGC0000720\n",
      "Scanning BGC0000721\n",
      "Scanning BGC0000722\n",
      "Scanning BGC0000723\n",
      "Scanning BGC0000724\n",
      "Scanning BGC0000725\n",
      "Scanning BGC0000726\n",
      "Scanning BGC0000727\n",
      "Scanning BGC0000728\n",
      "Scanning BGC0000729\n",
      "Scanning BGC0000730\n",
      "Scanning BGC0000731\n",
      "Scanning BGC0000732\n",
      "Scanning BGC0000733\n",
      "Scanning BGC0000734\n",
      "Scanning BGC0000735\n",
      "Scanning BGC0000736\n",
      "Scanning BGC0000737\n",
      "Scanning BGC0000738\n",
      "Scanning BGC0000739\n",
      "Scanning BGC0000740\n",
      "Scanning BGC0000741\n",
      "Scanning BGC0000742\n",
      "Scanning BGC0000743\n",
      "Scanning BGC0000744\n",
      "Scanning BGC0000745\n",
      "Scanning BGC0000746\n",
      "Scanning BGC0000747\n",
      "Scanning BGC0000748\n",
      "Scanning BGC0000749\n",
      "Scanning BGC0000750\n",
      "Scanning BGC0000751\n",
      "Scanning BGC0000752\n",
      "Scanning BGC0000753\n",
      "Scanning BGC0000754\n",
      "Scanning BGC0000755\n",
      "Scanning BGC0000756\n",
      "Scanning BGC0000757\n",
      "Scanning BGC0000758\n",
      "Scanning BGC0000759\n",
      "Scanning BGC0000760\n",
      "Scanning BGC0000761\n",
      "Scanning BGC0000762\n",
      "Scanning BGC0000763\n",
      "Scanning BGC0000764\n",
      "Scanning BGC0000765\n",
      "Scanning BGC0000766\n",
      "Scanning BGC0000767\n",
      "Scanning BGC0000768\n",
      "Scanning BGC0000769\n",
      "Scanning BGC0000770\n",
      "Scanning BGC0000771\n",
      "Scanning BGC0000772\n",
      "Scanning BGC0000773\n",
      "Scanning BGC0000774\n",
      "Scanning BGC0000775\n",
      "Scanning BGC0000776\n",
      "Scanning BGC0000777\n",
      "Scanning BGC0000778\n",
      "Scanning BGC0000779\n",
      "Scanning BGC0000780\n",
      "Scanning BGC0000781\n",
      "Scanning BGC0000782\n",
      "Scanning BGC0000783\n",
      "Scanning BGC0000784\n",
      "Scanning BGC0000785\n",
      "Scanning BGC0000786\n",
      "Scanning BGC0000787\n",
      "Scanning BGC0000788\n",
      "Scanning BGC0000789\n",
      "Scanning BGC0000790\n",
      "Scanning BGC0000791\n",
      "Scanning BGC0000792\n",
      "Scanning BGC0000793\n",
      "Scanning BGC0000794\n",
      "Scanning BGC0000795\n",
      "Scanning BGC0000796\n",
      "Scanning BGC0000797\n",
      "Scanning BGC0000798\n",
      "Scanning BGC0000799\n",
      "Scanning BGC0000800\n",
      "Scanning BGC0000801\n",
      "Scanning BGC0000802\n",
      "Scanning BGC0000803\n",
      "Scanning BGC0000804\n",
      "Scanning BGC0000805\n",
      "Scanning BGC0000806\n",
      "Scanning BGC0000807\n",
      "Scanning BGC0000808\n",
      "Scanning BGC0000809\n",
      "Scanning BGC0000810\n",
      "Scanning BGC0000811\n",
      "Scanning BGC0000812\n",
      "Scanning BGC0000813\n",
      "Scanning BGC0000814\n",
      "Scanning BGC0000815\n",
      "Scanning BGC0000816\n",
      "Scanning BGC0000817\n",
      "Scanning BGC0000818\n",
      "Scanning BGC0000819\n",
      "Scanning BGC0000820\n",
      "Scanning BGC0000821\n",
      "Scanning BGC0000822\n",
      "Scanning BGC0000823\n",
      "Scanning BGC0000824\n",
      "Scanning BGC0000825\n",
      "Scanning BGC0000826\n",
      "Scanning BGC0000827\n",
      "Scanning BGC0000828\n",
      "Scanning BGC0000829\n",
      "Scanning BGC0000830\n",
      "Scanning BGC0000831\n",
      "Scanning BGC0000832\n",
      "Scanning BGC0000833\n",
      "Scanning BGC0000834\n",
      "Scanning BGC0000835\n",
      "Scanning BGC0000836\n",
      "Scanning BGC0000837\n",
      "Scanning BGC0000838\n",
      "Scanning BGC0000839\n",
      "Scanning BGC0000840\n",
      "Scanning BGC0000841\n",
      "Scanning BGC0000842\n",
      "Scanning BGC0000843\n",
      "Scanning BGC0000844\n",
      "Scanning BGC0000845\n",
      "Scanning BGC0000846\n",
      "Scanning BGC0000847\n",
      "Scanning BGC0000848\n",
      "Scanning BGC0000849\n",
      "Scanning BGC0000850\n",
      "Scanning BGC0000851\n",
      "Scanning BGC0000852\n",
      "Scanning BGC0000853\n",
      "Scanning BGC0000854\n",
      "Scanning BGC0000855\n",
      "Scanning BGC0000856\n",
      "Scanning BGC0000857\n",
      "Scanning BGC0000858\n",
      "Scanning BGC0000859\n",
      "Scanning BGC0000860\n",
      "Scanning BGC0000861\n",
      "Scanning BGC0000862\n",
      "Scanning BGC0000863\n",
      "Scanning BGC0000864\n",
      "Scanning BGC0000865\n",
      "Scanning BGC0000866\n",
      "Scanning BGC0000867\n",
      "Scanning BGC0000868\n",
      "Scanning BGC0000869\n",
      "Scanning BGC0000870\n",
      "Scanning BGC0000871\n",
      "Scanning BGC0000872\n",
      "Scanning BGC0000873\n",
      "Scanning BGC0000874\n",
      "Scanning BGC0000875\n",
      "Scanning BGC0000876\n",
      "Scanning BGC0000877\n",
      "Scanning BGC0000878\n",
      "Scanning BGC0000879\n",
      "Scanning BGC0000880\n",
      "Scanning BGC0000881\n",
      "Scanning BGC0000882\n",
      "Scanning BGC0000883\n",
      "Scanning BGC0000884\n",
      "Scanning BGC0000885\n",
      "Scanning BGC0000886\n",
      "Scanning BGC0000887\n",
      "Scanning BGC0000888\n",
      "Scanning BGC0000889\n",
      "Scanning BGC0000890\n",
      "Scanning BGC0000891\n",
      "Scanning BGC0000892\n",
      "Scanning BGC0000893\n",
      "Scanning BGC0000894\n",
      "Scanning BGC0000895\n",
      "Scanning BGC0000896\n",
      "Scanning BGC0000897\n",
      "Scanning BGC0000898\n",
      "Scanning BGC0000899\n",
      "Scanning BGC0000900\n",
      "Scanning BGC0000901\n",
      "Scanning BGC0000902\n",
      "Scanning BGC0000903\n",
      "Scanning BGC0000904\n",
      "Scanning BGC0000905\n",
      "Scanning BGC0000906\n",
      "Scanning BGC0000907\n",
      "Scanning BGC0000908\n",
      "Scanning BGC0000909\n",
      "Scanning BGC0000910\n",
      "Scanning BGC0000911\n",
      "Scanning BGC0000912\n",
      "Scanning BGC0000913\n",
      "Scanning BGC0000914\n",
      "Scanning BGC0000915\n",
      "Scanning BGC0000916\n",
      "Scanning BGC0000917\n",
      "Scanning BGC0000918\n",
      "Scanning BGC0000919\n",
      "Scanning BGC0000920\n",
      "Scanning BGC0000921\n",
      "Scanning BGC0000922\n",
      "Scanning BGC0000923\n",
      "Scanning BGC0000924\n",
      "Scanning BGC0000925\n",
      "Scanning BGC0000926\n",
      "Scanning BGC0000927\n",
      "Scanning BGC0000928\n",
      "Scanning BGC0000929\n",
      "Scanning BGC0000930\n",
      "Scanning BGC0000931\n",
      "Scanning BGC0000932\n",
      "Scanning BGC0000933\n",
      "Scanning BGC0000934\n",
      "Scanning BGC0000935\n",
      "Scanning BGC0000936\n",
      "Scanning BGC0000937\n",
      "Scanning BGC0000938\n",
      "Scanning BGC0000939\n",
      "Scanning BGC0000940\n",
      "Scanning BGC0000941\n",
      "Scanning BGC0000942\n",
      "Scanning BGC0000943\n",
      "Scanning BGC0000944\n",
      "Scanning BGC0000945\n",
      "Scanning BGC0000946\n",
      "Scanning BGC0000947\n",
      "Scanning BGC0000948\n",
      "Scanning BGC0000949\n",
      "Scanning BGC0000950\n",
      "Scanning BGC0000951\n",
      "Scanning BGC0000952\n",
      "Scanning BGC0000953\n",
      "Scanning BGC0000954\n",
      "Scanning BGC0000955\n",
      "Scanning BGC0000956\n",
      "Scanning BGC0000957\n",
      "Scanning BGC0000958\n",
      "Scanning BGC0000959\n",
      "Scanning BGC0000960\n",
      "Scanning BGC0000961\n",
      "Scanning BGC0000962\n",
      "Scanning BGC0000963\n",
      "Scanning BGC0000964\n",
      "Scanning BGC0000965\n",
      "Scanning BGC0000966\n",
      "Scanning BGC0000967\n",
      "Scanning BGC0000968\n",
      "Scanning BGC0000969\n",
      "Scanning BGC0000970\n",
      "Scanning BGC0000971\n",
      "Scanning BGC0000972\n",
      "Scanning BGC0000973\n",
      "Scanning BGC0000974\n",
      "Scanning BGC0000975\n",
      "Scanning BGC0000976\n",
      "Scanning BGC0000977\n",
      "Scanning BGC0000978\n",
      "Scanning BGC0000979\n",
      "Scanning BGC0000980\n",
      "Scanning BGC0000981\n",
      "Scanning BGC0000982\n",
      "Scanning BGC0000983\n",
      "Scanning BGC0000984\n",
      "Scanning BGC0000985\n",
      "Scanning BGC0000986\n",
      "Scanning BGC0000988\n",
      "Scanning BGC0000989\n",
      "Scanning BGC0000990\n",
      "Scanning BGC0000991\n",
      "Scanning BGC0000992\n",
      "Scanning BGC0000993\n",
      "Scanning BGC0000994\n",
      "Scanning BGC0000995\n",
      "Scanning BGC0000996\n",
      "Scanning BGC0000997\n",
      "Scanning BGC0000998\n",
      "Scanning BGC0000999\n",
      "Scanning BGC0001000\n",
      "Scanning BGC0001001\n",
      "Scanning BGC0001002\n",
      "Scanning BGC0001003\n",
      "Scanning BGC0001004\n",
      "Scanning BGC0001005\n",
      "Scanning BGC0001006\n",
      "Scanning BGC0001007\n",
      "Scanning BGC0001008\n",
      "Scanning BGC0001009\n",
      "Scanning BGC0001010\n",
      "Scanning BGC0001011\n",
      "Scanning BGC0001012\n",
      "Scanning BGC0001013\n",
      "Scanning BGC0001014\n",
      "Scanning BGC0001015\n",
      "Scanning BGC0001016\n",
      "Scanning BGC0001017\n",
      "Scanning BGC0001018\n",
      "Scanning BGC0001019\n",
      "Scanning BGC0001020\n",
      "Scanning BGC0001021\n",
      "Scanning BGC0001022\n",
      "Scanning BGC0001023\n",
      "Scanning BGC0001024\n",
      "Scanning BGC0001025\n",
      "Scanning BGC0001026\n",
      "Scanning BGC0001027\n",
      "Scanning BGC0001028\n",
      "Scanning BGC0001029\n",
      "Scanning BGC0001030\n",
      "Scanning BGC0001031\n",
      "Scanning BGC0001032\n",
      "Scanning BGC0001033\n",
      "Scanning BGC0001034\n",
      "Scanning BGC0001035\n",
      "Scanning BGC0001036\n",
      "Scanning BGC0001037\n",
      "Scanning BGC0001038\n",
      "Scanning BGC0001039\n",
      "Scanning BGC0001040\n",
      "Scanning BGC0001041\n",
      "Scanning BGC0001042\n",
      "Scanning BGC0001043\n",
      "Scanning BGC0001044\n",
      "Scanning BGC0001045\n",
      "Scanning BGC0001046\n",
      "Scanning BGC0001047\n",
      "Scanning BGC0001048\n",
      "Scanning BGC0001049\n",
      "Scanning BGC0001050\n",
      "Scanning BGC0001051\n",
      "Scanning BGC0001052\n",
      "Scanning BGC0001053\n",
      "Scanning BGC0001054\n",
      "Scanning BGC0001055\n",
      "Scanning BGC0001056\n",
      "Scanning BGC0001057\n",
      "Scanning BGC0001058\n",
      "Scanning BGC0001059\n",
      "Scanning BGC0001060\n",
      "Scanning BGC0001061\n",
      "Scanning BGC0001062\n",
      "Scanning BGC0001063\n",
      "Scanning BGC0001064\n",
      "Scanning BGC0001065\n",
      "Scanning BGC0001066\n",
      "Scanning BGC0001067\n",
      "Scanning BGC0001068\n",
      "Scanning BGC0001069\n",
      "Scanning BGC0001070\n",
      "Scanning BGC0001071\n",
      "Scanning BGC0001072\n",
      "Scanning BGC0001073\n",
      "Scanning BGC0001074\n",
      "Scanning BGC0001075\n",
      "Scanning BGC0001076\n",
      "Scanning BGC0001077\n",
      "Scanning BGC0001078\n",
      "Scanning BGC0001079\n",
      "Scanning BGC0001080\n",
      "Scanning BGC0001081\n",
      "Scanning BGC0001082\n",
      "Scanning BGC0001083\n",
      "Scanning BGC0001084\n",
      "Scanning BGC0001085\n",
      "Scanning BGC0001086\n",
      "Scanning BGC0001087\n",
      "Scanning BGC0001088\n",
      "Scanning BGC0001089\n",
      "Scanning BGC0001090\n",
      "Scanning BGC0001091\n",
      "Scanning BGC0001092\n",
      "Scanning BGC0001093\n",
      "Scanning BGC0001094\n",
      "Scanning BGC0001095\n",
      "Scanning BGC0001096\n",
      "Scanning BGC0001098\n",
      "Scanning BGC0001099\n",
      "Scanning BGC0001100\n",
      "Scanning BGC0001101\n",
      "Scanning BGC0001102\n",
      "Scanning BGC0001103\n",
      "Scanning BGC0001104\n",
      "Scanning BGC0001105\n",
      "['629395', '2']\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001106\n",
      "Scanning BGC0001107\n",
      "Scanning BGC0001108\n",
      "['629395', '2']\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001109\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0001110\n",
      "Scanning BGC0001111\n",
      "Scanning BGC0001112\n",
      "Scanning BGC0001113\n",
      "Scanning BGC0001114\n",
      "Scanning BGC0001115\n",
      "Scanning BGC0001116\n",
      "Scanning BGC0001117\n",
      "Scanning BGC0001118\n",
      "Scanning BGC0001119\n",
      "Scanning BGC0001120\n",
      "Scanning BGC0001121\n",
      "Scanning BGC0001122\n",
      "Scanning BGC0001123\n",
      "Scanning BGC0001124\n",
      "Scanning BGC0001125\n",
      "Scanning BGC0001126\n",
      "Scanning BGC0001127\n",
      "Scanning BGC0001128\n",
      "Scanning BGC0001129\n",
      "Error...\n",
      "Error...\n",
      "Error...\n",
      "Error...\n",
      "Error...\n",
      "BGC0001129 Failed to download: MIBIG.BGC0001129.1\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001130\n",
      "Scanning BGC0001131\n",
      "Scanning BGC0001132\n",
      "Scanning BGC0001133\n",
      "Scanning BGC0001134\n",
      "Scanning BGC0001135\n",
      "Scanning BGC0001136\n",
      "Scanning BGC0001137\n",
      "Scanning BGC0001138\n",
      "Scanning BGC0001140\n",
      "Scanning BGC0001141\n",
      "Scanning BGC0001142\n",
      "Scanning BGC0001143\n",
      "Scanning BGC0001144\n",
      "Scanning BGC0001145\n",
      "[ripp] Can't find CDS feature for wp_004943468\n",
      "Scanning BGC0001146\n",
      "[ripp] Can't find CDS feature for aid54693\n",
      "Scanning BGC0001147\n",
      "Scanning BGC0001148\n",
      "Scanning BGC0001149\n",
      "Scanning BGC0001150\n",
      "Scanning BGC0001151\n",
      "Scanning BGC0001152\n",
      "Scanning BGC0001153\n",
      "Scanning BGC0001154\n",
      "Scanning BGC0001155\n",
      "[ripp] solved precursors BGC0001155\n",
      "Scanning BGC0001156\n",
      "Scanning BGC0001157\n",
      "[ripp] Can't find CDS feature for cbg72694,scab_56681\n",
      "Scanning BGC0001158\n",
      "Scanning BGC0001159\n",
      "Scanning BGC0001160\n",
      "Scanning BGC0001161\n",
      "Scanning BGC0001162\n",
      "Scanning BGC0001163\n",
      "Scanning BGC0001164\n",
      "Scanning BGC0001165\n",
      "Scanning BGC0001166\n",
      "Scanning BGC0001167\n",
      "[ripp] Wrong sequence information for afk79987.1\n",
      "[ripp] Wrong sequence information for afk79996.1\n",
      "[ripp] Wrong sequence information for afk79988.1\n",
      "Scanning BGC0001168\n",
      "Scanning BGC0001169\n",
      "Scanning BGC0001170\n",
      "[ripp] Can't find CDS feature for clospo_01168\n",
      "Scanning BGC0001171\n",
      "Scanning BGC0001172\n",
      "Scanning BGC0001173\n",
      "[ripp] Can't find CDS feature for bat_4209\n",
      "Scanning BGC0001174\n",
      "Scanning BGC0001176\n",
      "[ripp] Wrong sequence information for lp52_05076\n",
      "Scanning BGC0001177\n",
      "Scanning BGC0001178\n",
      "Scanning BGC0001179\n",
      "Scanning BGC0001180\n",
      "Scanning BGC0001181\n",
      "Scanning BGC0001182\n",
      "Scanning BGC0001183\n",
      "Scanning BGC0001184\n",
      "Scanning BGC0001185\n",
      "Scanning BGC0001186\n",
      "Scanning BGC0001187\n",
      "Scanning BGC0001188\n",
      "Scanning BGC0001189\n",
      "Scanning BGC0001190\n",
      "Scanning BGC0001191\n",
      "Scanning BGC0001192\n",
      "Scanning BGC0001193\n",
      "Scanning BGC0001194\n",
      "Scanning BGC0001195\n",
      "Scanning BGC0001196\n",
      "Scanning BGC0001197\n",
      "Scanning BGC0001198\n",
      "Scanning BGC0001199\n",
      "Scanning BGC0001200\n",
      "Scanning BGC0001201\n",
      "Scanning BGC0001202\n",
      "Scanning BGC0001203\n",
      "Scanning BGC0001204\n",
      "Scanning BGC0001205\n",
      "Scanning BGC0001206\n",
      "Scanning BGC0001207\n",
      "Scanning BGC0001208\n",
      "Scanning BGC0001209\n",
      "Scanning BGC0001210\n",
      "Scanning BGC0001211\n",
      "Scanning BGC0001212\n",
      "Scanning BGC0001213\n",
      "Scanning BGC0001214\n",
      "Scanning BGC0001215\n",
      "Scanning BGC0001216\n",
      "Scanning BGC0001217\n",
      "Scanning BGC0001218\n",
      "Scanning BGC0001219\n",
      "Scanning BGC0001220\n",
      "Scanning BGC0001221\n",
      "Scanning BGC0001222\n",
      "Scanning BGC0001223\n",
      "Scanning BGC0001224\n",
      "Scanning BGC0001225\n",
      "Scanning BGC0001226\n",
      "[ripp] Wrong sequence information for b446_00784\n",
      "[ripp] Can't find CDS feature for b446_34506\n",
      "Scanning BGC0001227\n",
      "Scanning BGC0001228\n",
      "Scanning BGC0001229\n",
      "[ripp] solved precursors BGC0001229\n",
      "Scanning BGC0001230\n",
      "Scanning BGC0001231\n",
      "Scanning BGC0001232\n",
      "Scanning BGC0001233\n",
      "Scanning BGC0001234\n",
      "Scanning BGC0001235\n",
      "Scanning BGC0001236\n",
      "Scanning BGC0001237\n",
      "Scanning BGC0001238\n",
      "Scanning BGC0001239\n",
      "Scanning BGC0001240\n",
      "Scanning BGC0001241\n",
      "Scanning BGC0001242\n",
      "Scanning BGC0001243\n",
      "Scanning BGC0001244\n",
      "Scanning BGC0001245\n",
      "Scanning BGC0001246\n",
      "Scanning BGC0001247\n",
      "Scanning BGC0001248\n",
      "Scanning BGC0001249\n",
      "Scanning BGC0001250\n",
      "Scanning BGC0001251\n",
      "Scanning BGC0001252\n",
      "Scanning BGC0001253\n",
      "Scanning BGC0001254\n",
      "Scanning BGC0001255\n",
      "Scanning BGC0001256\n",
      "Scanning BGC0001257\n",
      "Scanning BGC0001258\n",
      "Scanning BGC0001259\n",
      "Scanning BGC0001260\n",
      "Scanning BGC0001261\n",
      "Scanning BGC0001262\n",
      "Scanning BGC0001263\n",
      "Scanning BGC0001264\n",
      "Scanning BGC0001265\n",
      "Scanning BGC0001266\n",
      "Scanning BGC0001267\n",
      "Scanning BGC0001268\n",
      "Scanning BGC0001269\n",
      "Scanning BGC0001270\n",
      "Scanning BGC0001271\n",
      "Scanning BGC0001272\n",
      "Scanning BGC0001273\n",
      "Scanning BGC0001274\n",
      "Scanning BGC0001275\n",
      "Scanning BGC0001276\n",
      "Scanning BGC0001277\n",
      "Scanning BGC0001278\n",
      "Scanning BGC0001279\n",
      "Scanning BGC0001280\n",
      "Scanning BGC0001281\n",
      "Scanning BGC0001282\n",
      "Scanning BGC0001283\n",
      "Scanning BGC0001284\n",
      "Scanning BGC0001285\n",
      "Scanning BGC0001286\n",
      "Scanning BGC0001287\n",
      "Scanning BGC0001288\n",
      "Scanning BGC0001289\n",
      "Scanning BGC0001290\n",
      "Scanning BGC0001291\n",
      "Scanning BGC0001292\n",
      "Scanning BGC0001293\n",
      "Scanning BGC0001294\n",
      "Scanning BGC0001295\n",
      "Scanning BGC0001296\n",
      "Scanning BGC0001297\n",
      "Scanning BGC0001298\n",
      "Scanning BGC0001299\n",
      "Scanning BGC0001300\n",
      "Scanning BGC0001301\n",
      "Scanning BGC0001302\n",
      "Scanning BGC0001303\n",
      "Scanning BGC0001304\n",
      "Scanning BGC0001305\n",
      "Scanning BGC0001306\n",
      "Scanning BGC0001307\n",
      "Scanning BGC0001308\n",
      "Scanning BGC0001309\n",
      "Scanning BGC0001310\n",
      "Scanning BGC0001311\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111140\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111175\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111195\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111170\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111165\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111160\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111155\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111150\n",
      "[ripp] Wrong sequence information for rumflafd1_rs0111145\n",
      "Scanning BGC0001312\n",
      "Scanning BGC0001313\n",
      "Scanning BGC0001314\n",
      "Scanning BGC0001315\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001316\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001317\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001318\n",
      "Scanning BGC0001319\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001320\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001321\n",
      "Scanning BGC0001322\n",
      "Scanning BGC0001323\n",
      "Scanning BGC0001324\n",
      "Scanning BGC0001325\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001327\n",
      "Scanning BGC0001328\n",
      "Scanning BGC0001329\n",
      "Scanning BGC0001330\n",
      "Scanning BGC0001331\n",
      "Scanning BGC0001332\n",
      "Scanning BGC0001333\n",
      "Scanning BGC0001334\n",
      "Scanning BGC0001335\n",
      "Scanning BGC0001336\n",
      "Scanning BGC0001337\n",
      "Scanning BGC0001338\n",
      "Scanning BGC0001339\n",
      "Scanning BGC0001340\n",
      "Scanning BGC0001341\n",
      "Scanning BGC0001342\n",
      "Scanning BGC0001343\n",
      "Scanning BGC0001344\n",
      "Scanning BGC0001345\n",
      "Scanning BGC0001346\n",
      "Scanning BGC0001347\n",
      "Scanning BGC0001348\n",
      "Scanning BGC0001349\n",
      "Scanning BGC0001350\n",
      "Scanning BGC0001351\n",
      "Scanning BGC0001352\n",
      "Scanning BGC0001353\n",
      "Scanning BGC0001354\n",
      "Scanning BGC0001355\n",
      "Scanning BGC0001356\n",
      "Scanning BGC0001357\n",
      "Scanning BGC0001358\n",
      "Scanning BGC0001359\n",
      "Scanning BGC0001360\n",
      "Scanning BGC0001361\n",
      "Scanning BGC0001362\n",
      "Scanning BGC0001363\n",
      "Scanning BGC0001364\n",
      "Scanning BGC0001365\n",
      "Scanning BGC0001366\n",
      "Scanning BGC0001367\n",
      "Scanning BGC0001368\n",
      "Scanning BGC0001369\n",
      "Scanning BGC0001370\n",
      "Scanning BGC0001371\n",
      "Scanning BGC0001372\n",
      "Scanning BGC0001373\n",
      "Scanning BGC0001374\n",
      "Scanning BGC0001375\n",
      "Scanning BGC0001376\n",
      "Scanning BGC0001377\n",
      "Scanning BGC0001378\n",
      "Scanning BGC0001379\n",
      "Scanning BGC0001380\n",
      "Scanning BGC0001381\n",
      "Scanning BGC0001382\n",
      "Scanning BGC0001383\n",
      "Scanning BGC0001384\n",
      "Scanning BGC0001385\n",
      "Scanning BGC0001386\n",
      "Scanning BGC0001387\n",
      "Scanning BGC0001388\n",
      "Scanning BGC0001389\n",
      "Scanning BGC0001390\n",
      "Scanning BGC0001391\n",
      "Scanning BGC0001392\n",
      "Scanning BGC0001393\n",
      "Scanning BGC0001394\n",
      "Scanning BGC0001395\n",
      "Scanning BGC0001396\n",
      "Scanning BGC0001397\n",
      "Scanning BGC0001398\n",
      "Scanning BGC0001399\n",
      "Scanning BGC0001400\n",
      "Scanning BGC0001401\n",
      "Scanning BGC0001402\n",
      "Scanning BGC0001403\n",
      "Scanning BGC0001404\n",
      "Scanning BGC0001405\n",
      "Scanning BGC0001406\n",
      "Scanning BGC0001407\n",
      "Scanning BGC0001408\n",
      "Scanning BGC0001409\n",
      "Scanning BGC0001410\n",
      "Scanning BGC0001411\n",
      "Scanning BGC0001412\n",
      "Scanning BGC0001413\n",
      "Scanning BGC0001414\n",
      "Scanning BGC0001415\n",
      "Scanning BGC0001416\n",
      "Scanning BGC0001417\n",
      "Scanning BGC0001418\n",
      "Scanning BGC0001419\n",
      "Scanning BGC0001420\n",
      "Scanning BGC0001421\n",
      "Scanning BGC0001422\n",
      "Scanning BGC0001423\n",
      "Scanning BGC0001424\n",
      "Scanning BGC0001425\n",
      "Scanning BGC0001426\n",
      "Scanning BGC0001427\n",
      "Scanning BGC0001428\n",
      "Scanning BGC0001429\n",
      "Scanning BGC0001430\n",
      "Scanning BGC0001431\n",
      "Scanning BGC0001432\n",
      "Scanning BGC0001433\n",
      "Scanning BGC0001434\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanning BGC0001435\n",
      "Scanning BGC0001436\n",
      "Scanning BGC0001437\n",
      "Scanning BGC0001438\n",
      "Scanning BGC0001439\n",
      "Scanning BGC0001440\n",
      "Scanning BGC0001441\n",
      "Scanning BGC0001442\n",
      "Scanning BGC0001443\n",
      "Scanning BGC0001444\n",
      "Scanning BGC0001445\n",
      "Scanning BGC0001446\n",
      "Scanning BGC0001447\n",
      "Scanning BGC0001448\n",
      "Scanning BGC0001449\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001450\n",
      "Scanning BGC0001451\n",
      "Scanning BGC0001452\n",
      "Scanning BGC0001453\n",
      "Scanning BGC0001454\n",
      "Scanning BGC0001455\n",
      "Scanning BGC0001456\n",
      "Scanning BGC0001457\n",
      "Scanning BGC0001458\n",
      "Scanning BGC0001459\n",
      "Scanning BGC0001460\n",
      "Scanning BGC0001461\n",
      "Scanning BGC0001462\n",
      "Scanning BGC0001463\n",
      "Scanning BGC0001464\n",
      "Scanning BGC0001465\n",
      "Scanning BGC0001466\n",
      "Scanning BGC0001467\n",
      "Scanning BGC0001468\n",
      "Scanning BGC0001469\n",
      "Scanning BGC0001470\n",
      "Scanning BGC0001471\n",
      "[ripp] Wrong sequence information for wp_091614358.1\n",
      "Scanning BGC0001472\n",
      "Scanning BGC0001473\n",
      "[ripp] Wrong sequence information for pyc67624.1\n",
      "Scanning BGC0001474\n",
      "[ripp] solved precursors BGC0001474\n",
      "Scanning BGC0001475\n",
      "Scanning BGC0001476\n",
      "Scanning BGC0001477\n",
      "Scanning BGC0001478\n",
      "Scanning BGC0001479\n",
      "Scanning BGC0001480\n",
      "Scanning BGC0001481\n",
      "Scanning BGC0001483\n",
      "Scanning BGC0001484\n",
      "Scanning BGC0001485\n",
      "Scanning BGC0001486\n",
      "Scanning BGC0001487\n",
      "Scanning BGC0001488\n",
      "Scanning BGC0001489\n",
      "Scanning BGC0001490\n",
      "Scanning BGC0001491\n",
      "Scanning BGC0001492\n",
      "Scanning BGC0001493\n",
      "Scanning BGC0001494\n",
      "Scanning BGC0001495\n",
      "Scanning BGC0001496\n",
      "Scanning BGC0001497\n",
      "Scanning BGC0001498\n",
      "Scanning BGC0001499\n",
      "Scanning BGC0001500\n",
      "Scanning BGC0001501\n",
      "Scanning BGC0001502\n",
      "Scanning BGC0001503\n",
      "Scanning BGC0001504\n",
      "Scanning BGC0001505\n",
      "Scanning BGC0001506\n",
      "Scanning BGC0001507\n",
      "Scanning BGC0001508\n",
      "Scanning BGC0001509\n",
      "Scanning BGC0001510\n",
      "Scanning BGC0001511\n",
      "Scanning BGC0001512\n",
      "Scanning BGC0001513\n",
      "Scanning BGC0001514\n",
      "Scanning BGC0001515\n",
      "Scanning BGC0001516\n",
      "Scanning BGC0001517\n",
      "Scanning BGC0001518\n",
      "Scanning BGC0001519\n",
      "Scanning BGC0001520\n",
      "Scanning BGC0001521\n",
      "Scanning BGC0001522\n",
      "Scanning BGC0001523\n",
      "Scanning BGC0001524\n",
      "Scanning BGC0001525\n",
      "Scanning BGC0001526\n",
      "Scanning BGC0001527\n",
      "Scanning BGC0001528\n",
      "Scanning BGC0001529\n",
      "Scanning BGC0001530\n",
      "Scanning BGC0001531\n",
      "Scanning BGC0001532\n",
      "Scanning BGC0001533\n",
      "Scanning BGC0001534\n",
      "Scanning BGC0001535\n",
      "Scanning BGC0001536\n",
      "Scanning BGC0001537\n",
      "Scanning BGC0001538\n",
      "Scanning BGC0001539\n",
      "Scanning BGC0001540\n",
      "Scanning BGC0001541\n",
      "Scanning BGC0001542\n",
      "Scanning BGC0001543\n",
      "Scanning BGC0001544\n",
      "Scanning BGC0001545\n",
      "Scanning BGC0001546\n",
      "Scanning BGC0001547\n",
      "Scanning BGC0001548\n",
      "Scanning BGC0001549\n",
      "Scanning BGC0001550\n",
      "Scanning BGC0001551\n",
      "Scanning BGC0001552\n",
      "Scanning BGC0001553\n",
      "Scanning BGC0001554\n",
      "Scanning BGC0001555\n",
      "Scanning BGC0001556\n",
      "Scanning BGC0001557\n",
      "Scanning BGC0001558\n",
      "Scanning BGC0001559\n",
      "Scanning BGC0001560\n",
      "Scanning BGC0001561\n",
      "Scanning BGC0001562\n",
      "Scanning BGC0001563\n",
      "Scanning BGC0001564\n",
      "Scanning BGC0001565\n",
      "Scanning BGC0001566\n",
      "Scanning BGC0001567\n",
      "Scanning BGC0001568\n",
      "Scanning BGC0001569\n",
      "Scanning BGC0001570\n",
      "Scanning BGC0001571\n",
      "Scanning BGC0001572\n",
      "Scanning BGC0001573\n",
      "Scanning BGC0001574\n",
      "Scanning BGC0001575\n",
      "Scanning BGC0001576\n",
      "Scanning BGC0001577\n",
      "Scanning BGC0001578\n",
      "Scanning BGC0001579\n",
      "Scanning BGC0001580\n",
      "Scanning BGC0001581\n",
      "Scanning BGC0001582\n",
      "Scanning BGC0001583\n",
      "Scanning BGC0001584\n",
      "Scanning BGC0001585\n",
      "Scanning BGC0001586\n",
      "Scanning BGC0001587\n",
      "Scanning BGC0001588\n",
      "Scanning BGC0001589\n",
      "Scanning BGC0001590\n",
      "Scanning BGC0001591\n",
      "Scanning BGC0001592\n",
      "Scanning BGC0001593\n",
      "Scanning BGC0001594\n",
      "Scanning BGC0001595\n",
      "Scanning BGC0001596\n",
      "Scanning BGC0001597\n",
      "Scanning BGC0001598\n",
      "Scanning BGC0001599\n",
      "Scanning BGC0001600\n",
      "Scanning BGC0001601\n",
      "Scanning BGC0001602\n",
      "Scanning BGC0001603\n",
      "Scanning BGC0001604\n",
      "Scanning BGC0001605\n",
      "Scanning BGC0001606\n",
      "Scanning BGC0001607\n",
      "Scanning BGC0001608\n",
      "Scanning BGC0001609\n",
      "Scanning BGC0001610\n",
      "Scanning BGC0001611\n",
      "Scanning BGC0001612\n",
      "Scanning BGC0001613\n",
      "Scanning BGC0001614\n",
      "Scanning BGC0001615\n",
      "Scanning BGC0001616\n",
      "Scanning BGC0001617\n",
      "Scanning BGC0001618\n",
      "Scanning BGC0001619\n",
      "Scanning BGC0001620\n",
      "Scanning BGC0001621\n",
      "Scanning BGC0001622\n",
      "Scanning BGC0001623\n",
      "Scanning BGC0001624\n",
      "Scanning BGC0001625\n",
      "Scanning BGC0001626\n",
      "Scanning BGC0001627\n",
      "Scanning BGC0001628\n",
      "Scanning BGC0001629\n",
      "Scanning BGC0001630\n",
      "Scanning BGC0001631\n",
      "Scanning BGC0001632\n",
      "Scanning BGC0001633\n",
      "Scanning BGC0001634\n",
      "Scanning BGC0001635\n",
      "Scanning BGC0001636\n",
      "Scanning BGC0001637\n",
      "Scanning BGC0001638\n",
      "Scanning BGC0001639\n",
      "Scanning BGC0001640\n",
      "Scanning BGC0001641\n",
      "Scanning BGC0001642\n",
      "Scanning BGC0001643\n",
      "Scanning BGC0001644\n",
      "Scanning BGC0001645\n",
      "Scanning BGC0001646\n",
      "Scanning BGC0001647\n",
      "Scanning BGC0001648\n",
      "Scanning BGC0001649\n",
      "Scanning BGC0001650\n",
      "Scanning BGC0001651\n",
      "Scanning BGC0001652\n",
      "Scanning BGC0001653\n",
      "Scanning BGC0001654\n",
      "Scanning BGC0001655\n",
      "Scanning BGC0001656\n",
      "Scanning BGC0001657\n",
      "Scanning BGC0001658\n",
      "Scanning BGC0001659\n",
      "Scanning BGC0001660\n",
      "Scanning BGC0001661\n",
      "Scanning BGC0001662\n",
      "Scanning BGC0001663\n",
      "Scanning BGC0001664\n",
      "Scanning BGC0001665\n",
      "Scanning BGC0001666\n",
      "Scanning BGC0001667\n",
      "Scanning BGC0001668\n",
      "Scanning BGC0001669\n",
      "Scanning BGC0001670\n",
      "Scanning BGC0001671\n",
      "Scanning BGC0001672\n",
      "Scanning BGC0001673\n",
      "Scanning BGC0001674\n",
      "Scanning BGC0001675\n",
      "Scanning BGC0001676\n",
      "Scanning BGC0001677\n",
      "Scanning BGC0001678\n",
      "Scanning BGC0001679\n",
      "Scanning BGC0001680\n",
      "Scanning BGC0001681\n",
      "Scanning BGC0001682\n",
      "Scanning BGC0001683\n",
      "Scanning BGC0001684\n",
      "Scanning BGC0001685\n",
      "Scanning BGC0001686\n",
      "Scanning BGC0001687\n",
      "Scanning BGC0001688\n",
      "Scanning BGC0001689\n",
      "Scanning BGC0001690\n",
      "Scanning BGC0001691\n",
      "Scanning BGC0001692\n",
      "Scanning BGC0001693\n",
      "Scanning BGC0001694\n",
      "Scanning BGC0001695\n",
      "Scanning BGC0001696\n",
      "Scanning BGC0001697\n",
      "Scanning BGC0001698\n",
      "Scanning BGC0001699\n",
      "Scanning BGC0001700\n",
      "Scanning BGC0001701\n",
      "Scanning BGC0001702\n",
      "Scanning BGC0001703\n",
      "Scanning BGC0001704\n",
      "Scanning BGC0001705\n",
      "Scanning BGC0001706\n",
      "Scanning BGC0001707\n",
      "Scanning BGC0001708\n",
      "Scanning BGC0001709\n",
      "Scanning BGC0001710\n",
      "Scanning BGC0001711\n",
      "Scanning BGC0001712\n",
      "Scanning BGC0001713\n",
      "Scanning BGC0001714\n",
      "Scanning BGC0001715\n",
      "Scanning BGC0001716\n",
      "Scanning BGC0001717\n",
      "Scanning BGC0001718\n",
      "Scanning BGC0001719\n",
      "Scanning BGC0001720\n",
      "Scanning BGC0001721\n",
      "Scanning BGC0001722\n",
      "Scanning BGC0001723\n",
      "Scanning BGC0001724\n",
      "Scanning BGC0001725\n",
      "Scanning BGC0001726\n",
      "Scanning BGC0001727\n",
      "Scanning BGC0001728\n",
      "Scanning BGC0001729\n",
      "Scanning BGC0001730\n",
      "Scanning BGC0001731\n",
      "Scanning BGC0001732\n",
      "Scanning BGC0001733\n",
      "Scanning BGC0001734\n",
      "Scanning BGC0001735\n",
      "Scanning BGC0001736\n",
      "Scanning BGC0001737\n",
      "Scanning BGC0001738\n",
      "Scanning BGC0001739\n",
      "Scanning BGC0001740\n",
      "Scanning BGC0001741\n",
      "Scanning BGC0001742\n",
      "Scanning BGC0001743\n",
      "Scanning BGC0001745\n",
      "Scanning BGC0001746\n",
      "Scanning BGC0001747\n",
      "Scanning BGC0001748\n",
      "Scanning BGC0001749\n",
      "[]\n",
      "'ncbi_tax_id' is a required property\n",
      "'organism_name' is a required property\n",
      "Scanning BGC0001750\n",
      "Scanning BGC0001751\n",
      "Scanning BGC0001752\n",
      "Scanning BGC0001753\n",
      "Scanning BGC0001754\n",
      "Scanning BGC0001755\n",
      "Scanning BGC0001756\n",
      "Scanning BGC0001757\n",
      "Scanning BGC0001758\n",
      "Scanning BGC0001759\n",
      "Scanning BGC0001760\n",
      "Scanning BGC0001761\n",
      "Scanning BGC0001762\n",
      "Scanning BGC0001763\n",
      "Scanning BGC0001764\n",
      "Scanning BGC0001765\n",
      "Scanning BGC0001766\n",
      "Scanning BGC0001767\n",
      "Scanning BGC0001768\n",
      "Scanning BGC0001769\n",
      "Scanning BGC0001770\n",
      "Scanning BGC0001771\n",
      "Scanning BGC0001772\n",
      "Scanning BGC0001773\n",
      "Scanning BGC0001774\n",
      "Scanning BGC0001775\n",
      "Scanning BGC0001776\n",
      "Scanning BGC0001777\n",
      "Scanning BGC0001778\n",
      "Scanning BGC0001779\n",
      "Scanning BGC0001780\n",
      "Scanning BGC0001781\n",
      "Scanning BGC0001782\n",
      "Scanning BGC0001783\n",
      "Scanning BGC0001784\n",
      "Scanning BGC0001785\n",
      "Scanning BGC0001786\n",
      "Scanning BGC0001787\n",
      "Scanning BGC0001788\n",
      "Scanning BGC0001789\n",
      "Scanning BGC0001790\n",
      "Scanning BGC0001791\n",
      "Scanning BGC0001792\n",
      "Scanning BGC0001793\n",
      "Scanning BGC0001794\n",
      "Scanning BGC0001795\n",
      "Scanning BGC0001796\n",
      "Scanning BGC0001797\n",
      "Scanning BGC0001798\n",
      "Scanning BGC0001799\n",
      "Scanning BGC0001800\n",
      "Scanning BGC0001801\n",
      "Scanning BGC0001802\n",
      "Scanning BGC0001803\n",
      "Scanning BGC0001804\n",
      "Scanning BGC0001805\n",
      "Scanning BGC0001806\n",
      "Scanning BGC0001807\n",
      "Scanning BGC0001808\n",
      "Scanning BGC0001809\n",
      "Scanning BGC0001810\n",
      "Scanning BGC0001811\n",
      "Scanning BGC0001812\n",
      "Scanning BGC0001813\n",
      "Scanning BGC0001814\n",
      "Scanning BGC0001815\n",
      "Scanning BGC0001816\n",
      "Scanning BGC0001817\n",
      "Scanning BGC0001818\n",
      "Scanning BGC0001819\n",
      "Scanning BGC0001820\n",
      "Scanning BGC0001821\n",
      "Scanning BGC0001822\n",
      "Scanning BGC0001823\n",
      "Scanning BGC0001824\n",
      "Scanning BGC0001825\n",
      "Scanning BGC0001826\n",
      "Scanning BGC0001827\n",
      "Scanning BGC0001828\n",
      "Scanning BGC0001829\n",
      "Scanning BGC0001830\n",
      "All data fetched!\n"
     ]
    }
   ],
   "source": [
    "input_path = \"../../preprocessed/p6-json/\"\n",
    "output_folder = \"../../preprocessed/p7-json/\"\n",
    "\n",
    "if not path.exists(output_folder):\n",
    "    makedirs(output_folder)\n",
    "\n",
    "validator = None\n",
    "schema_props = {}\n",
    "with open(\"../../inputs/mibig_schema_phase_6.json\") as json_file:\n",
    "    schema_obj = json.load(json_file)\n",
    "    validator = Draft7Validator(schema_obj)\n",
    "    schema_props = fetch_props_new_schema(schema_obj, \"\", {})\n",
    "\n",
    "\n",
    "for json_path in sorted(glob.glob(path.join(input_path, \"BGC*.json\"))):\n",
    "    with open(json_path, \"r\") as json_file:\n",
    "        bgc_id = path.basename(json_path).split(\".\")[0]\n",
    "        data = json.load(json_file)\n",
    "        print(\"Scanning {}\".format(bgc_id))\n",
    "        check_gene_annotations(data)\n",
    "        validate_data(data, validator, schema_props)\n",
    "        with open(path.join(output_folder, \"{}.json\".format(bgc_id)), \"w\") as o:\n",
    "            o.write(json.dumps(data, indent=4, separators=(',', ': '), sort_keys=True))\n",
    "            \n",
    "print(\"All data fetched!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "retired_bgcs = {}\n",
    "todo_list = {}\n",
    "\n",
    "with open(\"../../preprocessed/reports/p7-removed_extra_genes.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\tgene_ids\\n\")\n",
    "    for bgc_id in removed_extra_genes:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \",\".join(removed_extra_genes[bgc_id])))\n",
    "        # uncommented: we don't need to retire BGCs because we trimmed out extra_genes\n",
    "        #if bgc_id not in retired_bgcs:\n",
    "        #    retired_bgcs[bgc_id] = set()\n",
    "        #retired_bgcs[bgc_id].add(\"extra_genes_removed\")\n",
    "    \n",
    "with open(\"../../preprocessed/reports/p7-removed_annotations.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\tgene_ids\\n\")\n",
    "    for bgc_id in removed_annotations:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \";\".join(removed_annotations[bgc_id])))\n",
    "        if bgc_id not in retired_bgcs:\n",
    "            retired_bgcs[bgc_id] = set()\n",
    "        retired_bgcs[bgc_id].add(\"annotations_removed\")\n",
    "    \n",
    "with open(\"../../preprocessed/reports/p7-removed_operons.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\tindexes\\n\")\n",
    "    for bgc_id in removed_operons:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \",\".join(removed_operons[bgc_id])))\n",
    "        if bgc_id not in retired_bgcs:\n",
    "            retired_bgcs[bgc_id] = set()\n",
    "        retired_bgcs[bgc_id].add(\"operons_removed\")\n",
    "\n",
    "with open(\"../../preprocessed/reports/p7-no_gbk.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\taccession\\n\")\n",
    "    for bgc_id in no_gbk:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, no_gbk[bgc_id]))\n",
    "        if bgc_id not in retired_bgcs:\n",
    "            retired_bgcs[bgc_id] = set()\n",
    "        retired_bgcs[bgc_id].add(\"no_gbk\")\n",
    "    \n",
    "with open(\"../../preprocessed/reports/p7-no_taxonomy.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\torganism\\n\")\n",
    "    for bgc_id in no_taxid:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, no_taxid[bgc_id]))\n",
    "        if bgc_id not in retired_bgcs:\n",
    "            retired_bgcs[bgc_id] = set()\n",
    "        retired_bgcs[bgc_id].add(\"no_taxonomy\")\n",
    "        \n",
    "        \n",
    "with open(\"../../preprocessed/reports/p7-retired_list.tsv\", \"w\") as o:\n",
    "    for bgc_id in retired_bgcs:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \";\".join(retired_bgcs[bgc_id])))\n",
    "    \n",
    "with open(\"../../preprocessed/reports/p7-added_mibig_genes.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\tgenes\\n\")\n",
    "    for bgc_id in added_mibig_genes:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \",\".join(added_mibig_genes[bgc_id])))\n",
    "\n",
    "with open(\"../../preprocessed/reports/p7-ripps_precursors_to_fix.tsv\", \"w\") as o:\n",
    "    for bgc_id in ripps_to_fix:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \",\".join([str(idx) for idx in ripps_to_fix[bgc_id]])))\n",
    "        if bgc_id not in todo_list:\n",
    "            todo_list[bgc_id] = {}\n",
    "        todo_list[bgc_id][\"ripp_precursors\"] = ripps_to_fix[bgc_id]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "### update fixed list ###\n",
    "with open(\"../../preprocessed/reports/p7-fixed_list.tsv\", \"w\") as o:\n",
    "    for bgc_id in fixed:\n",
    "        o.write(\"{}\\t{}\\n\".format(bgc_id, \";\".join(fixed[bgc_id])))\n",
    "        \n",
    "with open(\"../../preprocessed/reports/p7-todo_list.tsv\", \"w\") as o:\n",
    "    for bgc_id in todo_list:\n",
    "        for issue in todo_list[bgc_id]:\n",
    "            o.write(\"{}\\t{}\\t{}\\n\".format(bgc_id, issue, \";\".join([str(idx) for idx in todo_list[bgc_id][issue]])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../../preprocessed/p7-gene_counts_and_loci_length.tsv\", \"w\") as o:\n",
    "    o.write(\"bgc_id\\tnt_length\\tcds_count\\n\")\n",
    "    for bgc_id in nt_length_and_cds_count:\n",
    "        nt_length, gene_counts = nt_length_and_cds_count[bgc_id]\n",
    "        o.write(\"{}\\t{}\\t{}\\n\".format(bgc_id, nt_length, gene_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
